<!DOCTYPE html><html class="theme-next pisces use-motion" lang="zh-CN"><head><script></script><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=2"><meta name="theme-color" content="#222"><meta http-equiv="Cache-Control" content="no-transform"><meta http-equiv="Cache-Control" content="no-siteapp"><link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css"><link href="/css/main.css?v=6.3.0" rel="stylesheet" type="text/css"><link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=6.3.0"><link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=6.3.0"><link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=6.3.0"><link rel="mask-icon" href="/images/logo.svg?v=6.3.0" color="#222"><script type="text/javascript" id="hexo.configurations">var NexT=window.NexT||{},CONFIG={root:"/",scheme:"Pisces",version:"6.3.0",sidebar:{position:"left",display:"always",offset:12,b2t:!1,scrollpercent:!1,onmobile:!1},fancybox:!1,fastclick:!1,lazyload:!1,tabs:!0,motion:{enable:!0,async:!1,transition:{post_block:"fadeIn",post_header:"slideDownIn",post_body:"slideDownIn",coll_header:"slideLeftIn",sidebar:"slideUpIn"}},algolia:{applicationID:"",apiKey:"",indexName:"",hits:{per_page:10},labels:{input_placeholder:"Search for Posts",hits_empty:"We didn't find any results for the search: ${query}",hits_stats:"${hits} results found in ${time} ms"}}}</script><meta name="description" content="asm与基因表达1234567891011121314151617[18:40:00] qwzhou@login:~/practice/A549_BS/asmonGeneexpression :$ awk -v OFS=&quot;\t&quot; &apos;&amp;#123;if($7==&quot;-&quot;)&amp;#123;print $1,$5,&quot;-&quot;&amp;#125;else&amp;#123;print $1,$4,&quot;+&quot;&amp;#125;&amp;#125;&apos;"><meta name="keywords" content="methhap, project"><meta property="og:type" content="article"><meta property="og:title" content="project_methhap"><meta property="og:url" content="https://zhouqiangwei.github.io/2019/04/03/project-methhap/index.html"><meta property="og:site_name" content="Qiangwei&#39;s blog"><meta property="og:description" content="asm与基因表达1234567891011121314151617[18:40:00] qwzhou@login:~/practice/A549_BS/asmonGeneexpression :$ awk -v OFS=&quot;\t&quot; &apos;&amp;#123;if($7==&quot;-&quot;)&amp;#123;print $1,$5,&quot;-&quot;&amp;#125;else&amp;#123;print $1,$4,&quot;+&quot;&amp;#125;&amp;#125;&apos;"><meta property="og:locale" content="zh-CN"><meta property="og:updated_time" content="2019-05-11T10:50:30.645Z"><meta name="twitter:card" content="summary"><meta name="twitter:title" content="project_methhap"><meta name="twitter:description" content="asm与基因表达1234567891011121314151617[18:40:00] qwzhou@login:~/practice/A549_BS/asmonGeneexpression :$ awk -v OFS=&quot;\t&quot; &apos;&amp;#123;if($7==&quot;-&quot;)&amp;#123;print $1,$5,&quot;-&quot;&amp;#125;else&amp;#123;print $1,$4,&quot;+&quot;&amp;#125;&amp;#125;&apos;"><link rel="alternate" href="/atom.xml" title="Qiangwei's blog" type="application/atom+xml"><link rel="canonical" href="https://zhouqiangwei.github.io/2019/04/03/project-methhap/"><script type="text/javascript" id="page.configurations">CONFIG.page={sidebar:""}</script><title>project_methhap | Qiangwei's blog</title><noscript><style type="text/css">.sidebar-inner,.use-motion .brand,.use-motion .collection-title,.use-motion .comments,.use-motion .menu-item,.use-motion .motion-element,.use-motion .pagination,.use-motion .post-block,.use-motion .post-body,.use-motion .post-header{opacity:initial}.use-motion .logo,.use-motion .site-subtitle,.use-motion .site-title{opacity:initial;top:initial}.logo-line-after i{right:initial}</style></noscript></head><body itemscope itemtype="http://schema.org/WebPage" lang="zh-CN"><div class="container sidebar-position-left page-post-detail"><div class="headband"></div><header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader"><div class="header-inner"><div class="site-brand-wrapper"><div class="site-meta"><div class="custom-logo-site-title"><a href="/" class="brand" rel="start"><span class="logo-line-before"><i></i></span> <span class="site-title">Qiangwei's blog</span><span class="logo-line-after"><i></i></span></a></div><p class="site-subtitle">coding and learning</p></div><div class="site-nav-toggle"> <button aria-label="切换导航栏"><span class="btn-bar"></span><span class="btn-bar"></span><span class="btn-bar"></span></button></div></div><nav class="site-nav"><ul id="menu" class="menu"><li class="menu-item menu-item-home"><a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i><br>首页</a></li><li class="menu-item menu-item-tags"><a href="/tags/" rel="section"><i class="menu-item-icon fa fa-fw fa-tags"></i><br>标签</a></li><li class="menu-item menu-item-categories"><a href="/categories/" rel="section"><i class="menu-item-icon fa fa-fw fa-th"></i><br>分类</a></li><li class="menu-item menu-item-archives"><a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i><br>归档</a></li><li class="menu-item menu-item-about"><a href="/about/" rel="section"><i class="menu-item-icon fa fa-fw fa-user"></i><br>关于</a></li><li class="menu-item menu-item-guestbook"><a href="/guestbook/" rel="section"><i class="menu-item-icon fa fa-fw fa-newspaper-o"></i><br>留言板</a></li><li class="menu-item menu-item-search"><a href="javascript:;" class="popup-trigger"><i class="menu-item-icon fa fa-search fa-fw"></i><br>搜索</a></li></ul><div class="site-search"><div class="popup search-popup local-search-popup"><div class="local-search-header clearfix"><span class="search-icon"><i class="fa fa-search"></i></span><span class="popup-btn-close"><i class="fa fa-times-circle"></i></span><div class="local-search-input-wrapper"> <input autocomplete="off" placeholder="搜索..." spellcheck="false" type="text" id="local-search-input"></div></div><div id="local-search-result"></div></div></div></nav></div></header><main id="main" class="main"><div class="main-inner"><div class="content-wrap"><div id="content" class="content"><div id="posts" class="posts-expand"><article class="post post-type-normal" itemscope itemtype="http://schema.org/Article"><div class="post-block"><link itemprop="mainEntityOfPage" href="https://zhouqiangwei.github.io/2019/04/03/project-methhap/"><span hidden itemprop="author" itemscope itemtype="http://schema.org/Person"><meta itemprop="name" content="Qiangwei"><meta itemprop="description" content="顺风不浪 逆风不怂"><meta itemprop="image" content="/images/avatar.jpg"></span><span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization"><meta itemprop="name" content="Qiangwei's blog"></span><header class="post-header"><h1 class="post-title" itemprop="name headline">project_methhap</h1><div class="post-meta"><span class="post-time"><span class="post-meta-item-icon"><i class="fa fa-calendar-o"></i></span> <span class="post-meta-item-text">发表于</span> <time title="创建时间：2019-04-03 15:36:47" itemprop="dateCreated datePublished" datetime="2019-04-03T15:36:47+08:00">2019-04-03</time> <span class="post-meta-divider">|</span><span class="post-meta-item-icon"><i class="fa fa-calendar-check-o"></i></span> <span class="post-meta-item-text">更新于</span> <time title="修改时间：2019-05-11 18:50:30" itemprop="dateModified" datetime="2019-05-11T18:50:30+08:00">2019-05-11</time></span> <span class="post-category"><span class="post-meta-divider">|</span><span class="post-meta-item-icon"><i class="fa fa-folder-o"></i></span> <span class="post-meta-item-text">分类于</span> <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/project/" itemprop="url" rel="index"><span itemprop="name">project</span></a></span></span> <span class="post-comments-count"><span class="post-meta-divider">|</span><span class="post-meta-item-icon"><i class="fa fa-comment-o"></i></span> <a href="/2019/04/03/project-methhap/#comments" itemprop="discussionUrl"><span class="post-meta-item-text">评论数：</span><span class="post-comments-count valine-comment-count" data-xid="/2019/04/03/project-methhap/" itemprop="commentCount"></span></a></span> <span id="/2019/04/03/project-methhap/" class="leancloud_visitors" data-flag-title="project_methhap"><span class="post-meta-divider">|</span><span class="post-meta-item-icon"><i class="fa fa-eye"></i></span> <span class="post-meta-item-text">阅读次数：</span><span class="leancloud-visitors-count"></span></span></div></header><div class="post-body" itemprop="articleBody"><script src="/assets/js/APlayer.min.js"></script><h3 id="asm与基因表达"><a href="#asm与基因表达" class="headerlink" title="asm与基因表达"></a>asm与基因表达</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">[18:40:00] qwzhou@login:~/practice/A549_BS/asmonGeneexpression :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;if($7=="-")&#123;print $1,$5,"-"&#125;else&#123;print $1,$4,"+"&#125;&#125;'</span> gex.big0Head.gtf &gt; gex.big0Head.TSS.ped3</span><br><span class="line">[18:40:50] qwzhou@login:~/practice/A549_BS/asmonGeneexpression :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;if($7=="-")&#123;print $1,$5,"-"&#125;else&#123;print $1,$4,"+"&#125;&#125;'</span> gex.big0Tail.gtf &gt; gex.big0Tail.TSS.ped3</span><br><span class="line">[18:41:16] qwzhou@login:~/practice/A549_BS/asmonGeneexpression :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;if($7=="-")&#123;print $1,$5,"-"&#125;else&#123;print $1,$4,"+"&#125;&#125;'</span> gex.eq0.gtf &gt; gex.eq0.TSS.ped3</span><br><span class="line"></span><br><span class="line"><span class="comment">###区分两类来看</span></span><br><span class="line">[18:53:56] qwzhou@login:~/practice/A549_BS/asmonGeneexpression :</span><br><span class="line">$ cat gex.big0Head.gtf gex.big0Tail.gtf &gt; gex.expression.gtf</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;if($7=="-")&#123;print $1,$5,"-"&#125;else&#123;print $1,$4,"+"&#125;&#125;'</span> gex.expression.gtf &gt; gex.expression.TSS.ped3</span><br><span class="line"></span><br><span class="line"><span class="comment">##获得富集情况</span></span><br><span class="line">[19:01:54] qwzhou@login:~/practice/A549_BS/asmonGeneexpression :</span><br><span class="line">$ awk -v num=21238 <span class="string">'&#123;printf $1;for(i=2;i&lt;=NF;i++)&#123;printf "\t"100*$i/num&#125;&#125;'</span> A549.asmonExpressionGene.Methy.1.txt &gt; A549.asmonExpressionGene.Methy.1.txt.Aver</span><br><span class="line">[19:02:45] qwzhou@login:~/practice/A549_BS/asmonGeneexpression :</span><br><span class="line">$ awk -v num=37005 <span class="string">'&#123;printf $1;for(i=2;i&lt;=NF;i++)&#123;printf "\t"100*$i/num&#125;&#125;'</span> A549.asmonUnGene.Methy.1.txt &gt; A549.asmonUnGene.Methy.1.txt.Aver</span><br></pre></td></tr></table></figure><h3 id="asm的分布"><a href="#asm的分布" class="headerlink" title="asm的分布"></a>asm的分布</h3><h4 id="ASM在基因上的分布"><a href="#ASM在基因上的分布" class="headerlink" title="ASM在基因上的分布"></a>ASM在基因上的分布</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"> awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;if($7=="+")&#123;print $1,$4,$7&#125;else&#123;print $1,$5,$7&#125;&#125;'</span> ~/practice/Genome/hg38/gene.gtf &gt; gene.ped3</span><br><span class="line">[11:29:32] qwzhou@login:~/practice/hepG2/WGBS :</span><br><span class="line">$ ~/software_devp/methyhaplo/ASMannoSites -o HepG2.asmonTSS -G ~/practice/Genome/hg38/batmeth2-chr/hg38.chr.fa --ped ~/practice/Genome/hg38/gene.ped3 -d 5000 -ap hepG2.newasm.plus.txt -an hepG2.newasm.neg.txt</span><br><span class="line"><span class="comment"># etc....</span></span><br></pre></td></tr></table></figure><h4 id="ASM在exon上的分布"><a href="#ASM在exon上的分布" class="headerlink" title="ASM在exon上的分布"></a>ASM在exon上的分布</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$3=="exon"&#123;if($7=="+")&#123;print $1,$4,$7&#125;else&#123;print $1,$5,$7&#125;&#125;'</span> ~/practice/Genome/hg38/Homo_sapiens.GRCh38.90.addchr.gtf &gt; ~/practice/Genome/hg38/exon.ped3</span><br><span class="line">[12:58:48] qwzhou@login:~/practice/hepG2/WGBS :</span><br><span class="line">$ ~/software_devp/methyhaplo/ASMannoSites -o HepG2.asmonExon -G ~/practice/Genome/hg38/batmeth2-chr/hg38.chr.fa -b ~/practice/Genome/hg38/exon.ped3 -d 5000 -ap hepG2.newasm.plus.txt -an hepG2.newasm.neg.txt</span><br><span class="line">...</span><br></pre></td></tr></table></figure><h4 id="获取intron区域"><a href="#获取intron区域" class="headerlink" title="获取intron区域"></a>获取intron区域</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">[11:47:07] qwzhou@login:~/practice/Genome/hg38 :</span><br><span class="line">$ awk <span class="string">'BEGIN&#123;FS="\t| |;";OFS="\t"&#125;$3=="exon"&#123;print $1,$4,$5,$10,$22,$7&#125;'</span> Homo_sapiens.GRCh38.90.addchr.gtf &gt; exon.bed</span><br><span class="line"></span><br><span class="line">[11:47:43] qwzhou@login:~/practice/Genome/hg38 :</span><br><span class="line">$ awk <span class="string">'BEGIN&#123;FS="\t| |;";OFS="\t"&#125;$3=="transcript"&#123;print $1,$4,$5,$10,$22,$7&#125;'</span> Homo_sapiens.GRCh38.90.addchr.gtf &gt; transcript.bed</span><br><span class="line"></span><br><span class="line"><span class="comment">## 排序</span></span><br><span class="line">[11:48:00] qwzhou@login:~/practice/Genome/hg38 :</span><br><span class="line">$ bedtools sort -i transcript.bed &gt; transcript.sorted.bed</span><br><span class="line">[11:48:35] qwzhou@login:~/practice/Genome/hg38 :</span><br><span class="line">$ bedtools sort -i exon.bed &gt; exon.sorted.bed</span><br><span class="line"></span><br><span class="line"><span class="comment">## 合并区域</span></span><br><span class="line">[11:48:45] qwzhou@login:~/practice/Genome/hg38 :</span><br><span class="line">$ bedtools merge -i transcript.sorted.bed -s -c 4,5,6 -o distinct &gt; transcript.sorted.merged.bed</span><br><span class="line">[11:48:58] qwzhou@login:~/practice/Genome/hg38 :</span><br><span class="line">$ bedtools merge -i exon.sorted.bed -s -c 4,5,6 -o distinct &gt; exon.sorted.merged.bed</span><br><span class="line"></span><br><span class="line"><span class="comment">## 相减获得intron区域</span></span><br><span class="line"><span class="comment">## bedtools subtract -s -a transcript.sorted.merged.bed -b exon.sorted.merged.bed &gt; intron.bed</span></span><br><span class="line">生成的结果与transcript文件一样的，失败。。。</span><br></pre></td></tr></table></figure><h4 id="ASM在intron上的分布"><a href="#ASM在intron上的分布" class="headerlink" title="ASM在intron上的分布"></a>ASM在intron上的分布</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">intron提取失败，用外显子的尾端来测试</span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$3=="exon"&#123;if($7=="+")&#123;print $1,$5,$7&#125;else&#123;print $1,$4,$7&#125;&#125;'</span> ~/practice/Genome/hg38/Homo_sapiens.GRCh38.90.addchr.gtf &gt; ~/practice/Genome/hg38/exonEnd.ped3</span><br><span class="line">[12:58:48] qwzhou@login:~/practice/hepG2/WGBS :</span><br><span class="line">$ ~/software_devp/methyhaplo/ASMannoSites -o HepG2.asmonExonend -G ~/practice/Genome/hg38/batmeth2-chr/hg38.chr.fa -b ~/practice/Genome/hg38/exonEnd.ped3 -d 5000 -ap hepG2.newasm.plus.txt -an hepG2.newasm.neg.txt</span><br><span class="line">etc...</span><br></pre></td></tr></table></figure><h3 id="获取基因间区"><a href="#获取基因间区" class="headerlink" title="获取基因间区"></a>获取基因间区</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">'BEGIN&#123;OFS="\t";&#125; $3=="gene" &#123;print $1,$4-1,$5&#125;'</span> Homo_sapiens.GRCh38.90.addchr.gtf | sortBed | complementBed -i stdin -g hg38.chr.fa.len.sort &gt; hg38_intergenic2.bed</span><br></pre></td></tr></table></figure><h3 id="获取基因间区不包含promoter"><a href="#获取基因间区不包含promoter" class="headerlink" title="获取基因间区不包含promoter"></a>获取基因间区不包含promoter</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">[13:58:26] qwzhou@login:~/practice/Genome/hg38 :</span><br><span class="line">$ awk <span class="string">'BEGIN&#123;OFS="\t";&#125;ARGIND==1 &#123;a[$1]=$2&#125; ARGIND==2 &amp;&amp; $3=="gene" &#123;if($7=="+")&#123;if($4&gt;2000)&#123;s=$4-2000&#125;else&#123;s=$4&#125;;print $1,s,$5&#125;else&#123;if($5+2000 &lt;a[$1])&#123;s=$5+2000&#125;else&#123;s=$5&#125;;print $1,$4,s&#125;&#125;'</span> hg38.chr.fa.len.sort Homo_sapiens.GRCh38.90.addchr.gtf | sortBed | complementBed -i stdin -g hg38.chr.fa.len.sort &gt; hg38_intergenic.noPromoter.bed</span><br></pre></td></tr></table></figure><h3 id="查看asm-site在不同功能元件的分布"><a href="#查看asm-site在不同功能元件的分布" class="headerlink" title="查看asm site在不同功能元件的分布"></a>查看asm site在不同功能元件的分布</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">[13:40:39] qwzhou@login:~/practice/hepG2/WGBS :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/^*/&#123;print $1,$2"\n"$1,$3&#125;'</span> hepG2.newasm.plus.txt  hepG2.newasm.neg.txt | sort -k1,1 -k2,2n |uniq &gt; hepG2.asm.loci.txt</span><br><span class="line"></span><br><span class="line">[13:43:19] qwzhou@login:~/practice/hepG2/WGBS :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;print $1,$2,$2&#125;'</span> hepG2.asm.loci.txt |bedtools intersect -a -  -b ~/practice/Genome/hg38/hg38_intergenic2.bed -wo | awk <span class="string">'&#123;a[$1" "$2]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line">16317</span><br><span class="line">[13:49:51] qwzhou@login:~/practice/hepG2/WGBS :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;print $1,$2,$2&#125;'</span> hepG2.asm.loci.txt |bedtools intersect -a -  -b ~/practice/Genome/hg38/hg38_intron.bed -wo | awk <span class="string">'&#123;a[$1" "$2]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line">20607</span><br><span class="line">[13:50:07] qwzhou@login:~/practice/hepG2/WGBS :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;print $1,$2,$2&#125;'</span> hepG2.asm.loci.txt |bedtools intersect -a -  -b ~/practice/Genome/hg38/hg38_exon_merged.bed -wo | awk <span class="string">'&#123;a[$1" "$2]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line">6047</span><br><span class="line">[14:02:25] qwzhou@login:~/practice/hepG2/WGBS :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;print $1,$2,$2&#125;'</span> hepG2.asm.loci.txt |bedtools intersect -a -  -b ~/practice/Genome/hg38/hg38_intergenic.noPromoter.bed -wo | awk <span class="string">'&#123;a[$1" "$2]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line">13802</span><br></pre></td></tr></table></figure><h3 id="amrfinder"><a href="#amrfinder" class="headerlink" title="amrfinder"></a>amrfinder</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">'ARGIND==1 &amp;&amp; $3=="+" &amp;&amp; $5&lt;0.05 &amp;&amp; sqrt($NF*$NF)&gt;0.6'</span> BS.cvi.ler.dmc | awk -v lp=0,lc=<span class="string">"N"</span>,a=<span class="string">""</span>,an=0 <span class="string">'</span></span><br><span class="line"><span class="string">&#123;</span></span><br><span class="line"><span class="string">    if(lc!=$1)&#123;</span></span><br><span class="line"><span class="string">        if(an&gt;3)&#123;</span></span><br><span class="line"><span class="string">            print a;</span></span><br><span class="line"><span class="string">        &#125;</span></span><br><span class="line"><span class="string">        lp=$2;lc=$1;a=$0;an=1</span></span><br><span class="line"><span class="string">    &#125;else&#123;</span></span><br><span class="line"><span class="string">        if(lp&gt;=$2-150)&#123;</span></span><br><span class="line"><span class="string">            a=a"\n"$0;an++</span></span><br><span class="line"><span class="string">        &#125;else if(an&gt;3)&#123;</span></span><br><span class="line"><span class="string">            print a;a="";</span></span><br><span class="line"><span class="string">            a=$0;an=1;</span></span><br><span class="line"><span class="string">        &#125;else&#123;a=$0;an=1;&#125;</span></span><br><span class="line"><span class="string">        lp=$2;lc=$1;</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">&#125;</span></span><br><span class="line"><span class="string">'</span> |less</span><br><span class="line"></span><br><span class="line">[11:25:46] qwzhou@mn02:~/project/methhap/tair :</span><br><span class="line">$ awk <span class="string">'ARGIND==1 &amp;&amp; $3=="+" &amp;&amp; $5&lt;0.05 &amp;&amp; sqrt($NF*$NF)&gt;0.6'</span> BS.cvi.ler.dmc | awk -v lp=0,lc=<span class="string">"N"</span>,a=<span class="string">""</span>,an=0 <span class="string">'</span></span><br><span class="line"><span class="string">&#123;</span></span><br><span class="line"><span class="string">    if(lc!=$1)&#123;</span></span><br><span class="line"><span class="string">        if(an&gt;3)&#123;</span></span><br><span class="line"><span class="string">            print a;</span></span><br><span class="line"><span class="string">        &#125;</span></span><br><span class="line"><span class="string">        lp=$2;lc=$1;a=$0;an=1</span></span><br><span class="line"><span class="string">    &#125;else&#123;</span></span><br><span class="line"><span class="string">        if(lp&gt;=$2-150)&#123;</span></span><br><span class="line"><span class="string">            a=a"\n"$0;an++</span></span><br><span class="line"><span class="string">        &#125;else if(an&gt;3)&#123;</span></span><br><span class="line"><span class="string">            print a;a="";</span></span><br><span class="line"><span class="string">            a=$0;an=1;</span></span><br><span class="line"><span class="string">        &#125;else&#123;a=$0;an=1;&#125;</span></span><br><span class="line"><span class="string">        lp=$2;lc=$1;</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">&#125;</span></span><br><span class="line"><span class="string">'</span> | awk <span class="string">'ARGIND==1&#123;a[$1" "$2]&#125;ARGIND==2 &amp;&amp; $5&lt;0.01&#123;if($1" "($2+1) in a)&#123;print $0&#125;&#125;'</span> - BS-cviler.allelicmeth |wc -l</span><br><span class="line">627</span><br><span class="line">[11:25:58] qwzhou@mn02:~/project/methhap/tair :</span><br><span class="line">$ awk <span class="string">'$5&lt;0.01'</span> BS-cviler.allelicmeth |wc -l</span><br><span class="line">1562</span><br><span class="line"></span><br><span class="line">[11:27:58] qwzhou@mn02:~/project/methhap/tair :</span><br><span class="line">$ awk <span class="string">'ARGIND==1 &amp;&amp; $3=="+" &amp;&amp; $5&lt;0.05 &amp;&amp; sqrt($NF*$NF)&gt;0.6'</span> BS.cvi.ler.dmc | awk -v lp=0,lc=<span class="string">"N"</span>,a=<span class="string">""</span>,an=0 <span class="string">'</span></span><br><span class="line"><span class="string">&#123;</span></span><br><span class="line"><span class="string">    if(lc!=$1)&#123;</span></span><br><span class="line"><span class="string">        if(an&gt;3)&#123;</span></span><br><span class="line"><span class="string">            print a;</span></span><br><span class="line"><span class="string">        &#125;</span></span><br><span class="line"><span class="string">        lp=$2;lc=$1;a=$0;an=1</span></span><br><span class="line"><span class="string">    &#125;else&#123;</span></span><br><span class="line"><span class="string">        if(lp&gt;=$2-150)&#123;</span></span><br><span class="line"><span class="string">            a=a"\n"$0;an++</span></span><br><span class="line"><span class="string">        &#125;else if(an&gt;3)&#123;</span></span><br><span class="line"><span class="string">            print a;a="";</span></span><br><span class="line"><span class="string">            a=$0;an=1;</span></span><br><span class="line"><span class="string">        &#125;else&#123;a=$0;an=1;&#125;</span></span><br><span class="line"><span class="string">        lp=$2;lc=$1;</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">&#125;</span></span><br><span class="line"><span class="string">'</span> | awk <span class="string">'ARGIND==1&#123;a[$1" "$2]&#125;ARGIND==2 &amp;&amp; $1!~/^*/ &amp;&amp; $1=="1"&#123;if($1" "$2 in a || $1" "$3 in a)&#123;print $0&#125;&#125;'</span> - BS-cviler.asm.cg.plus.txt |wc -l</span><br><span class="line">1097</span><br><span class="line">[11:28:19] qwzhou@mn02:~/project/methhap/tair :</span><br><span class="line">$ awk <span class="string">'$1!~/^*/ &amp;&amp; $1=="1"'</span> BS-cviler.asm.cg.plus.txt |wc -l</span><br><span class="line">1466</span><br><span class="line"></span><br><span class="line">[16:56:14] qwzhou@mn02:~/project/methhap/tair :</span><br><span class="line">$ awk <span class="string">'$5&lt;0.01'</span> BS-cviler.allelicmeth | awk <span class="string">'&#123;print $1"\t"$2"\t"$2&#125;'</span> | bedtools intersect -a - -b BS-cviler.amr -wo | awk <span class="string">'&#123;a[$1" "$2]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line">1261</span><br><span class="line">[16:56:30] qwzhou@mn02:~/project/methhap/tair :</span><br><span class="line">$ awk <span class="string">'$5&lt;0.01'</span> BS-cviler.allelicmeth | wc -l</span><br><span class="line">1562</span><br><span class="line">[16:56:38] qwzhou@mn02:~/project/methhap/tair :</span><br><span class="line">$ awk <span class="string">'ARGIND==1 &amp;&amp; $5&lt;0.01&#123;a[$1" "$2+1]&#125;ARGIND==2 &amp;&amp; $1!~/^*/&#123;if($1" "$2 in a || $1" "$3 in a)&#123;print $0&#125;&#125;'</span> BS-cviler.allelicmeth BS-cviler.asm.cg.plus.txt |wc -l</span><br><span class="line">123</span><br></pre></td></tr></table></figure><h4 id="prec"><a href="#prec" class="headerlink" title="prec"></a>prec</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">[17:22:50] qwzhou@c06n06:~/project/methhap/bsChip/PrEC :</span><br><span class="line">$ samtools view -h PrEC.sort.bam.nosnp.meth.bam.G1.bam | awk <span class="string">'$5&gt;=20'</span> | wc -l</span><br><span class="line">2558109</span><br><span class="line">[17:23:02] qwzhou@c06n06:~/project/methhap/bsChip/PrEC :</span><br><span class="line">$ samtools view -h PrEC.sort.bam.nosnp.meth.bam.G2.bam | awk <span class="string">'$5&gt;=20'</span> | wc -l</span><br><span class="line">1787038</span><br><span class="line">[17:23:10] qwzhou@c06n06:~/project/methhap/bsChip/PrEC :</span><br><span class="line">$ samtools view -h PrEC.sort.bam.snp.bam.G1.bam | awk <span class="string">'$5&gt;=20'</span> | wc -l</span><br><span class="line">1869200</span><br><span class="line">[17:23:26] qwzhou@c06n06:~/project/methhap/bsChip/PrEC :</span><br><span class="line">$ samtools view -h PrEC.sort.bam.snp.bam.G2.bam | awk <span class="string">'$5&gt;=20'</span> | wc -l</span><br><span class="line">787527</span><br><span class="line">[17:23:35] qwzhou@c06n06:~/project/methhap/bsChip/PrEC :</span><br><span class="line">$ samtools view -h PrEC.sort.bam | awk <span class="string">'$5&gt;=20'</span> | wc -l</span><br><span class="line">126312219</span><br></pre></td></tr></table></figure><h3 id="A549-Oct4-检测"><a href="#A549-Oct4-检测" class="headerlink" title="A549 Oct4 检测"></a>A549 Oct4 检测</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">[15:41:20] qwzhou@mn02:~/project/methhap/Oct4 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$7=="PASS" &amp;&amp; $NF~/0\/1/&#123;print $3,$1,$2,"1",$4"/"$5&#125;'</span> ../WGBS/A549/wgbs.new.replace.mdups.snp.filtered.sort.vcf  &gt; ../A549.snp.SNPsplit.vcf</span><br></pre></td></tr></table></figure><h4 id="需要先建立mask基因组索引"><a href="#需要先建立mask基因组索引" class="headerlink" title="需要先建立mask基因组索引"></a>需要先建立mask基因组索引</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^#/&#123;print $0&#125;$1!~/^#/&#123;if($7=="PASS")&#123;print $1,$2,$3,$4,$5,$6,$7,$8,$9":FI",$10":1"&#125;else&#123;print $1,$2,$3,$4,$5,$6,$7,$8,$9":FI",$10":0"&#125;&#125;'</span> /public/home/qwzhou/project/methhap/wgbs.new.replace.mdups.snp.filtered.sort.vcf &gt; wgbs.snp.filtered.SNPsplit.vcf</span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^#/&#123;print $0&#125;$1!~/^#/ &amp;&amp; $NF~/0\/1/&#123;sub(/0\/1/,"1/1",$10);print $1,$2,$3,$4,$5,$6,$7,$8,$9,$10&#125;'</span> wgbs.snp.filtered.SNPsplit.vcf  &gt; wgbs.replace.snp.filtered.SNPsplit.replaceheterowithhomo.vcf</span><br><span class="line">[18:57:42] qwzhou@c01n12:~/project/methhap/Oct4 :</span><br><span class="line">$ SNPsplit_genome_preparation --vcf_file ./bowtie2index/wgbs.replace.snp.filtered.SNPsplit.replaceheterowithhomo.vcf --reference_genome bowtie2index --strain A549</span><br><span class="line"></span><br><span class="line">~~~ SNPsplit_genome_preparation --vcf_file /public/home/qwzhou/project/methhap/wgbs.new.replace.mdups.snp.filtered.sort.vcf --reference_genome bowtie2index --strain A549</span><br><span class="line">~~~</span><br><span class="line">[18:07:52] qwzhou@c01n12:~/project/methhap/Oct4/A549_N-masked :</span><br><span class="line">$ cat ./*fa &gt; hg38.N-masked.fa</span><br><span class="line">rm chrchr*</span><br><span class="line">bowtie2-build hg38.N-masked.fa hg38.N-masked.fa</span><br></pre></td></tr></table></figure><h3 id="bschip"><a href="#bschip" class="headerlink" title="bschip"></a>bschip</h3><h4 id="hct116"><a href="#hct116" class="headerlink" title="hct116"></a>hct116</h4><p>首先建立mask-genome索引，完成比对，这里使用bwa<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br></pre></td><td class="code"><pre><span class="line">[19:01:30] qwzhou@c02n06:~/project/methhap/bsChip/hct116/bwamethindex :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^#/&#123;print $0&#125;$1!~/^#/&#123;if($7=="PASS")&#123;print $1,$2,$3,$4,$5,$6,$7,$8,$9":FI",$10":1"&#125;else&#123;print $1,$2,$3,$4,$5,$6,$7,$8,$9":FI",$10":0"&#125;&#125;'</span> /public/home/qwzhou/project/methhap/bsChip/hct116_bs/hct116_bs/GSE118030/BS-Seq/GSM3317488/GSM3317488.vcf &gt;  hct116.snp.filtered.SNPsplit.vcf</span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^#/&#123;print $0&#125;$1!~/^#/ &amp;&amp; $NF~/0\/1/&#123;sub(/0\/1/,"1/1",$10);print $1,$2,$3,$4,$5,$6,$7,$8,$9,$10&#125;'</span> hct116.snp.filtered.SNPsplit.vcf  &gt; hct116.snp.filtered.SNPsplit.replaceheterowithhomo.vcf</span><br><span class="line"><span class="comment">## 拷贝vcf基因组信息到vcf文件</span></span><br><span class="line"><span class="comment">##reference=&lt;ID=null,Source=file:/public/home/qwzhou/practice/Genome/hg38/hg38.chr.fa&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chrM,length=16569,assembly=null,md5=c68f52674c9fb33aef52dcf399755519,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr1,length=248956422,assembly=null,md5=2648ae1bacce4ec4b6cf337dcae37816,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr2,length=242193529,assembly=null,md5=4bb4f82880a14111eb7327169ffb729b,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr3,length=198295559,assembly=null,md5=a48af509898d3736ba95dc0912c0b461,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr4,length=190214555,assembly=null,md5=3210fecf1eb92d5489da4346b3fddc6e,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr5,length=181538259,assembly=null,md5=f7f05fb7ceea78cbc32ce652c540ff2d,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr6,length=170805979,assembly=null,md5=6a48dfa97e854e3c6f186c8ff973f7dd,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr7,length=159345973,assembly=null,md5=94eef2b96fd5a7c8db162c8c74378039,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr8,length=145138636,assembly=null,md5=c67955b5f7815a9a1edfaa15893d3616,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr9,length=138394717,assembly=null,md5=addd2795560986b7491c40b1faa3978a,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr10,length=133797422,assembly=null,md5=907112d17fcb73bcab1ed1c72b97ce68,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr11,length=135086622,assembly=null,md5=1511375dc2dd1b633af8cf439ae90cec,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr12,length=133275309,assembly=null,md5=e81e16d3f44337034695a29b97708fce,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr13,length=114364328,assembly=null,md5=17dab79b963ccd8e7377cef59a54fe1c,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr14,length=107043718,assembly=null,md5=acbd9552c059d9b403e75ed26c1ce5bc,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr15,length=101991189,assembly=null,md5=f036bd11158407596ca6bf3581454706,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr16,length=90338345,assembly=null,md5=24e7cabfba3548a2bb4dff582b9ee870,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr17,length=83257441,assembly=null,md5=a8499ca51d6fb77332c2d242923994eb,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr18,length=80373285,assembly=null,md5=11eeaa801f6b0e2e36a1138616b8ee9a,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr19,length=58617616,assembly=null,md5=b0eba2c7bb5c953d1e06a508b5e487de,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr20,length=64444167,assembly=null,md5=b18e6c531b0bd70e949a7fc20859cb01,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr21,length=46709983,assembly=null,md5=2f45a3455007b7e271509161e52954a9,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chr22,length=50818468,assembly=null,md5=221733a2a15e2de66d33e73d126c5109,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chrY,length=57227415,assembly=null,md5=b2b7e6369564d89059e763cd6e736837,species=null&gt;</span></span><br><span class="line"><span class="comment">##contig=&lt;ID=chrX,length=156040895,assembly=null,md5=49527016a48497d9d1cbd8e4a9049bd3,species=null&gt;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">###修改strain为hct116</span></span><br><span class="line">[19:09:12] qwzhou@c02n06:~/project/methhap/bsChip/hct116 :</span><br><span class="line">$ SNPsplit_genome_preparation --vcf_file ./bwamethindex/hct116.snp.filtered.SNPsplit.replaceheterowithhomo.vcf --reference_genome bwamethindex/ --strain hct116</span><br><span class="line"></span><br><span class="line">[19:09:43] qwzhou@c02n06:~/project/methhap/bsChip/hct116/hct116_N-masked :</span><br><span class="line">$ cat ./*fa &gt; hg38.N-masked.fa</span><br><span class="line">[19:09:50] qwzhou@c02n06:~/project/methhap/bsChip/hct116/hct116_N-masked :</span><br><span class="line">$ rm chrchr*</span><br><span class="line">[19:20:14] qwzhou@c02n06:~/project/methhap/bsChip/hct116/hct116_N-masked :</span><br><span class="line">$ python ~/software/bwa-meth-master/bwameth.py index hg38.N-masked.fa</span><br><span class="line"></span><br><span class="line">[15:58:18] qwzhou@mn02:~/project/methhap/bsChip/hct116 :</span><br><span class="line">$ awk <span class="string">'$1~/^#/ || $NF~/0\/1/'</span> bwamethindex/hct116.snp.filtered.SNPsplit.vcf &gt; bwamethindex/hct116.snp.filtered.hetero.vcf</span><br><span class="line"></span><br><span class="line">[12:05:40] qwzhou@c03n06:~/project/methhap/bsChip/hct116 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$6&gt;=10 &amp;&amp; ($7&gt;=0.2 &amp;&amp; $7&lt;=0.8)&#123;if($3=="+")&#123;print $1,$2,".\tC\tT\t600\tPASS\t.\tGT\t0/1"&#125;else&#123;print $1,$2,".\tG\tA\t600\tPASS\t.\tGT\t0/1"&#125;&#125;'</span> ../hct116_bs/hct116_bs/GSE118030/BS-Seq/GSM3317488/GSM3317488.methratio.txt &gt; hct116.mr.vcf</span><br><span class="line">awk <span class="string">'ARGIND==1 &amp;&amp; $1~/^#/&#123;print $0&#125;ARGIND==2&#123;print $0&#125;'</span> bwamethindex/hct116.snp.filtered.hetero.vcf hct116.mr.vcf &gt; hct116.mr.withheader.vcf</span><br></pre></td></tr></table></figure><p></p><h4 id="PrEC"><a href="#PrEC" class="headerlink" title="PrEC"></a>PrEC</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">[15:54:49] qwzhou@mn02:~/project/methhap/bsChip/PrEC :</span><br><span class="line">$ awk <span class="string">'$1~/^#/ || $NF~/0\/1/'</span> prepareGenome/PrEC.snp.filtered.SNPsplit.vcf &gt; prepareGenome/PrEC.snp.filtered.hetero.vcf</span><br><span class="line">[19:28:09] qwzhou@mn02:~/project/methhap/bsChip/PrEC :</span><br><span class="line">$ SNPsplit_genome_preparation --vcf_file ./prepareGenome/PrEC.snp.filtered.SNPsplit.replaceheterowithhomo.vcf --reference_genome prepareGenome/ --strain PrEC</span><br><span class="line">[19:28:32] qwzhou@mn02:~/project/methhap/bsChip/PrEC/PrEC_N-masked :</span><br><span class="line">$ cat ./*fa &gt; hg38.N-masked.fa</span><br><span class="line">rm chrchr*</span><br><span class="line">[19:29:52] qwzhou@c02n05:~/project/methhap/bsChip/PrEC/PrEC_N-masked :</span><br><span class="line">$ python ~/software/bwa-meth-master/bwameth.py index hg38.N-masked.fa</span><br></pre></td></tr></table></figure><h4 id="LNCaP"><a href="#LNCaP" class="headerlink" title="LNCaP"></a>LNCaP</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">[14:35:15] qwzhou@mn02:~/project/methhap/bsChip/LNCaP :</span><br><span class="line">$ awk <span class="string">'$1~/^#/ || $NF~/0\/1/'</span> preGenome/LNCaP.snp.filtered.SNPsplit.vcf &gt; preGenome/LNCaP.snp.filtered.hetero.vcf</span><br><span class="line">[19:35:27] qwzhou@mn02:~/project/methhap/bsChip/LNCaP :</span><br><span class="line">$ SNPsplit_genome_preparation --vcf_file ./preGenome/LNCaP.snp.filtered.SNPsplit.replaceheterowithhomo.vcf --reference_genome preGenome/ --strain LNCaP</span><br><span class="line">[19:37:29] qwzhou@mn02:~/project/methhap/bsChip/LNCaP/LNCaP_N-masked :</span><br><span class="line">$ cat ./*fa &gt; hg38.N-masked.fa</span><br><span class="line">[19:38:05] qwzhou@mn02:~/project/methhap/bsChip/LNCaP/LNCaP_N-masked :</span><br><span class="line">$ rm chrchr*</span><br><span class="line">python ~/software/bwa-meth-master/bwameth.py index hg38.N-masked.fa</span><br></pre></td></tr></table></figure><h4 id="运算"><a href="#运算" class="headerlink" title="运算"></a>运算</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br></pre></td><td class="code"><pre><span class="line">[15:56:56] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$7=="PASS" &amp;&amp; $10~/0\/1/&#123;print $3,$1,$2,"1",$4"/"$5&#125;'</span> prostate/prostate_bs/GSE86833/BS-Seq/GSM2309190/GSM2309190.vcf &gt; prostate/PrEC.snp.snpsplit.vcf</span><br><span class="line">[15:57:39] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$7=="PASS" &amp;&amp; $10~/0\/1/&#123;print $3,$1,$2,"1",$4"/"$5&#125;'</span> prostate/prostate_bs/GSE86833/BS-Seq/GSM2309189/GSM2309189.vcf &gt; prostate/LNCaP.snp.snpsplit.vcf</span><br><span class="line"></span><br><span class="line">[15:59:49] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$3=="+" &amp;&amp; $6&gt;=10 &amp;&amp; ($7&gt;=0.2 &amp;&amp; $7&lt;=0.8)&#123;print $1,$2,".\tC\tT\t600\tPASS\t.\tGT\t0/1"&#125;'</span> prostate/prostate_bs/GSE86833/BS-Seq/GSM2309190/GSM2309190.methratio.txt &gt; prostate/PrEC.mr.p.vcf</span><br><span class="line">[16:07:44] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$3=="-" &amp;&amp; $6&gt;=10 &amp;&amp; ($7&gt;=0.2 &amp;&amp; $7&lt;=0.8)&#123;print $1,$2,".\tG\tA\t600\tPASS\t.\tGT\t0/1"&#125;'</span>  prostate/prostate_bs/GSE86833/BS-Seq/GSM2309190/GSM2309190.methratio.txt &gt; prostate/PrEC.mr.n.vcf</span><br><span class="line">[16:14:26] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$3=="+" &amp;&amp; $6&gt;=10 &amp;&amp; ($7&gt;=0.2 &amp;&amp; $7&lt;=0.8)&#123;print $1,$2,".\tC\tT\t600\tPASS\t.\tGT\t0/1"&#125;'</span> prostate/prostate_bs/GSE86833/BS-Seq/GSM2309189/GSM2309189.methratio.txt &gt; prostate/LNCaP.mr.p.vcf</span><br><span class="line">[16:19:15] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$3=="-" &amp;&amp; $6&gt;=10 &amp;&amp; ($7&gt;=0.2 &amp;&amp; $7&lt;=0.8)&#123;print $1,$2,".\tG\tA\t600\tPASS\t.\tGT\t0/1"&#125;'</span>  prostate/prostate_bs/GSE86833/BS-Seq/GSM2309189/GSM2309189.methratio.txt &gt; prostate/LNCaP.mr.n.vcf</span><br><span class="line">[16:49:49] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;print $3,$1,$2,"1",$4"/"$5&#125;'</span> prostate/PrEC.mr.p.vcf &gt; prostate/PrEC.mr.p.snpsplit.vcf</span><br><span class="line">[16:50:04] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;print $3,$1,$2,"1",$4"/"$5&#125;'</span> prostate/PrEC.mr.n.vcf &gt; prostate/PrEC.mr.n.snpsplit.vcf</span><br><span class="line">[16:52:28] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;print $3,$1,$2,"1",$4"/"$5&#125;'</span> prostate/LNCaP.mr.n.vcf &gt; prostate/LNCaP.mr.n.snpsplit.vcf</span><br><span class="line">[16:52:49] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;print $3,$1,$2,"1",$4"/"$5&#125;'</span> prostate/LNCaP.mr.p.vcf &gt; prostate/LNCaP.mr.p.snpsplit.vcf</span><br><span class="line"></span><br><span class="line">[16:54:49] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ cat prostate/PrEC.snp.snpsplit.vcf prostate/PrEC.mr.p.snpsplit.vcf | sort -k2,2 -k3,3n &gt; prostate/PrEC.mrsnp.p.snpsplit.vcf</span><br><span class="line">[16:55:13] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ cat prostate/PrEC.snp.snpsplit.vcf prostate/PrEC.mr.n.snpsplit.vcf | sort -k2,2 -k3,3n &gt; prostate/PrEC.mrsnp.n.snpsplit.vcf</span><br><span class="line">[16:55:40] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ cat prostate/LNCaP.snp.snpsplit.vcf prostate/LNCaP.mr.p.snpsplit.vcf | sort -k2,2 -k3,3n &gt; prostate/LNCaP.mrsnp.p.snpsplit.vcf</span><br><span class="line">[16:56:14] qwzhou@mn02:~/project/methhap/bsChip :</span><br><span class="line">$ cat prostate/LNCaP.snp.snpsplit.vcf prostate/LNCaP.mr.n.snpsplit.vcf | sort -k2,2 -k3,3n &gt; prostate/LNCaP.mrsnp.n.snpsplit.vcf</span><br><span class="line"></span><br><span class="line"><span class="comment">##hct116</span></span><br><span class="line">[17:09:39] qwzhou@mn02:~/project/methhap/bsChip/hct116_bs :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$7=="PASS" &amp;&amp; $10~/0\/1/&#123;print $3,$1,$2,"1",$4"/"$5&#125;'</span> hct116_bs/GSE118030/BS-Seq/GSM3317488/GSM3317488.vcf &gt; hct116.snp.snpsplit.vcf</span><br><span class="line">[17:10:49] qwzhou@mn02:~/project/methhap/bsChip/hct116_bs :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$3=="+" &amp;&amp; $6&gt;=10 &amp;&amp; ($7&gt;=0.2 &amp;&amp; $7&lt;=0.8)&#123;print $1,$2,".\tC\tT\t600\tPASS\t.\tGT\t0/1"&#125;'</span> hct116_bs/GSE118030/BS-Seq/GSM3317488/GSM3317488.methratio.txt &gt; hct116.mr.p.vcf</span><br><span class="line">[17:21:42] qwzhou@mn02:~/project/methhap/bsChip/hct116_bs :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$3=="-" &amp;&amp; $6&gt;=10 &amp;&amp; ($7&gt;=0.2 &amp;&amp; $7&lt;=0.8)&#123;print $1,$2,".\tG\tA\t600\tPASS\t.\tGT\t0/1"&#125;'</span> hct116_bs/GSE118030/BS-Seq/GSM3317488/GSM3317488.methratio.txt &gt; hct116.mr.n.vcf</span><br></pre></td></tr></table></figure><h3 id="6个细胞系的ASMG-GO分析"><a href="#6个细胞系的ASMG-GO分析" class="headerlink" title="6个细胞系的ASMG GO分析"></a>6个细胞系的ASMG GO分析</h3><p>首先获取asmg<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">[10:45:18] qwzhou@mn02:~/project/methhap/WGBS :</span><br><span class="line">bedtools intersect -a ~/project/Genome/hg38/gene.id.bed -b A549/A549.newasm.bed -wo | cut -f1-6 | uniq &gt; A549.asmg.bed</span><br><span class="line">bedtools intersect -a ~/project/Genome/hg38/gene.id.bed -b imr90/imr90.asm.bed -wo | cut -f1-6 | uniq &gt; imr90.asmg.bed</span><br><span class="line">bedtools intersect -a ~/project/Genome/hg38/gene.id.bed -b gm12878/GM12878.asm.bed -wo | cut -f1-6 | uniq &gt; gm12878.asmg.bed</span><br><span class="line">bedtools intersect -a ~/project/Genome/hg38/gene.id.bed -b hepG2/hepG2.newasm.bed -wo | cut -f1-6 | uniq &gt; hepG2.asmg.bed</span><br><span class="line">bedtools intersect -a ~/project/Genome/hg38/gene.id.bed -b hues64/hues64.asm.bed -wo | cut -f1-6 | uniq &gt; hues64.asmg.bed</span><br><span class="line">bedtools intersect -a ~/project/Genome/hg38/gene.id.bed -b k562/k562.asm.bed -wo | cut -f1-6 | uniq &gt; k562.asmg.bed</span><br></pre></td></tr></table></figure><p></p><h3 id="meth单独的单倍型长度"><a href="#meth单独的单倍型长度" class="headerlink" title="meth单独的单倍型长度"></a>meth单独的单倍型长度</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### hepG2</span></span><br><span class="line">[12:21:40] qwzhou@mn02:~/project/methhap/WGBS/hepG2 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> hepG2.mr.hapcut2.p.haplo.txt hepG2.mr.hapcut2.n.haplo.txt  | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | sort -k1,1 -k2,2n -k3,3n | bedtools merge &gt; hepG2.mr.haplo.merge.txt</span><br><span class="line">[12:43:04] qwzhou@mn02:~/project/methhap/WGBS/hepG2 :</span><br><span class="line">$ awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span> hepG2.mr.haplo.merge.txt</span><br><span class="line">683 14502 104522</span><br><span class="line">[12:43:28] qwzhou@mn02:~/project/methhap/WGBS/hepG2 :</span><br><span class="line">$ awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span> hepG2.mr.haplo.merge.txt</span><br><span class="line">314959538 1667917 188.834</span><br><span class="line"></span><br><span class="line"><span class="comment">### A549</span></span><br><span class="line"><span class="comment">## meth + snp</span></span><br><span class="line">[15:44:13] qwzhou@mn02:~/project/methhap/WGBS/A549 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> A549.snpwithmr.p.hapcut2.haplo.txt A549.snpwithmr.n.hapcut2.haplo.txt | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | sort -k1,1 -k2,2n -k3,3n | bedtools merge &gt; A549.snpwithmr.haplo.merge.txt</span><br><span class="line">[15:46:01] qwzhou@mn02:~/project/methhap/WGBS/A549 :</span><br><span class="line">$ awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span> A549.snpwithmr.haplo.merge.txt</span><br><span class="line">670171648 2045102 327.696</span><br><span class="line"></span><br><span class="line">[19:21:07] qwzhou@mn02:~/project/methhap/WGBS/A549 :</span><br><span class="line">$ awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span> A549.snpwithmr.haplo.merge.txt</span><br><span class="line">14140 95721 301276</span><br><span class="line"><span class="comment">## meth</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> A549.mr.hapcut2.p.haplo.txt A549.mr.hapcut2.n.haplo.txt  | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | sort -k1,1 -k2,2n -k3,3n | bedtools merge &gt; A549.mr.haplo.merge.txt</span><br><span class="line">awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span> A549.mr.haplo.merge.txt</span><br><span class="line">5381 58183 234785</span><br><span class="line">awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span> A549.mr.haplo.merge.txt</span><br><span class="line">525438628 1873522 280.455</span><br><span class="line"></span><br><span class="line"><span class="comment">## snp</span></span><br><span class="line">[19:19:05] qwzhou@mn02:~/project/methhap/WGBS/A549 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> A549.snv.hapcut2.haplo.txt | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span></span><br><span class="line">78669651 295069 266.614</span><br><span class="line"></span><br><span class="line">[19:19:59] qwzhou@mn02:~/project/methhap/WGBS/A549 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> A549.snv.hapcut2.haplo.txt | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span></span><br><span class="line">749 5944 30007</span><br><span class="line"></span><br><span class="line"><span class="comment">## + hic</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'BEGIN&#123;a=0;b=0;chr=""&#125;&#123;</span></span><br><span class="line"><span class="string">     if($1~/^B/)&#123;</span></span><br><span class="line"><span class="string">         if(b&gt;a)&#123;print chr,a,b;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">         else&#123;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">     else if(a==0)&#123;</span></span><br><span class="line"><span class="string">         chr=$1;a=$2;b=$2</span></span><br><span class="line"><span class="string">     &#125;else&#123;</span></span><br><span class="line"><span class="string">         if($2&gt;b+3000)&#123;</span></span><br><span class="line"><span class="string">             if(b&gt;a)&#123;print chr,a,b; a=$2;b=$2;chr=$1&#125;</span></span><br><span class="line"><span class="string">         &#125;else if($2&gt;b)&#123;</span></span><br><span class="line"><span class="string">             b=$2</span></span><br><span class="line"><span class="string">         &#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(b&gt;a)&#123;print chr,a,b&#125;&#125;'</span> A549.snpwithmr.merge.hic.haplo.txt | awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span></span><br><span class="line">883475472 2284995 386.642</span><br><span class="line"></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'BEGIN&#123;a=0;b=0;chr=""&#125;&#123;</span></span><br><span class="line"><span class="string">     if($1~/^B/)&#123;</span></span><br><span class="line"><span class="string">         if(b&gt;a)&#123;print chr,a,b;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">         else&#123;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">     else if(a==0)&#123;</span></span><br><span class="line"><span class="string">         chr=$1;a=$2;b=$2</span></span><br><span class="line"><span class="string">     &#125;else&#123;</span></span><br><span class="line"><span class="string">         if($2&gt;b+3000)&#123;</span></span><br><span class="line"><span class="string">             if(b&gt;a)&#123;print chr,a,b; a=$2;b=$2;chr=$1&#125;</span></span><br><span class="line"><span class="string">         &#125;else if($2&gt;b)&#123;</span></span><br><span class="line"><span class="string">             b=$2</span></span><br><span class="line"><span class="string">         &#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(b&gt;a)&#123;print chr,a,b&#125;&#125;'</span> A549.snpwithmr.merge.hic.haplo.txt | awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span></span><br><span class="line">28314 85952 298869</span><br></pre></td></tr></table></figure><h3 id="A549-覆盖repeat情况"><a href="#A549-覆盖repeat情况" class="headerlink" title="A549 覆盖repeat情况"></a>A549 覆盖repeat情况</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### + hic</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'BEGIN&#123;a=0;b=0;chr=""&#125;&#123;</span></span><br><span class="line"><span class="string">     if($1~/^B/)&#123;</span></span><br><span class="line"><span class="string">         if(b&gt;a)&#123;print chr,a,b;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">         else&#123;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">     else if(a==0)&#123;</span></span><br><span class="line"><span class="string">         chr=$1;a=$2;b=$2</span></span><br><span class="line"><span class="string">     &#125;else&#123;</span></span><br><span class="line"><span class="string">         if($2&gt;b+3000)&#123;</span></span><br><span class="line"><span class="string">             if(b&gt;a)&#123;print chr,a,b; a=$2;b=$2;chr=$1&#125;</span></span><br><span class="line"><span class="string">         &#125;else if($2&gt;b)&#123;</span></span><br><span class="line"><span class="string">             b=$2</span></span><br><span class="line"><span class="string">         &#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(b&gt;a)&#123;print chr,a,b&#125;&#125;'</span> A549.snpwithmr.merge.hic.haplo.txt | bedtools intersect -a - -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line">2326516</span><br><span class="line"><span class="comment">## meth snp</span></span><br><span class="line">bedtools intersect -a A549.snpwithmr.haplo.merge.txt -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line">2207366</span><br><span class="line"><span class="comment">## meth</span></span><br><span class="line">bedtools intersect -a A549.mr.haplo.merge.txt -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line">1845620</span><br><span class="line"><span class="comment">## snp</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> A549.snv.hapcut2.haplo.txt | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | bedtools intersect -a - -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br></pre></td></tr></table></figure><h3 id="统计IMR90、hepG2、K562的单倍型结果"><a href="#统计IMR90、hepG2、K562的单倍型结果" class="headerlink" title="统计IMR90、hepG2、K562的单倍型结果"></a>统计IMR90、hepG2、K562的单倍型结果</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 150</span></span><br><span class="line">scp ./wgbs.replace.mdups.snp.filtered.sort.vcf qwzhou@211.69.141.142:/public/home/qwzhou/project/methhap/WGBS/imr90</span><br><span class="line">scp ./run.hapcut2snv.sh qwzhou@211.69.141.142:/public/home/qwzhou/project/methhap/WGBS/imr90</span><br><span class="line">scp ./run.hapcut2snv.sh qwzhou@211.69.141.142:/public/home/qwzhou/project/methhap/WGBS/imr90</span><br><span class="line"><span class="comment">## 142</span></span><br><span class="line">[15:38:15] qwzhou@mn02:~/project/methhap/WGBS/imr90 :</span><br><span class="line">$ bsub &lt; run.lsf</span><br><span class="line"></span><br><span class="line"><span class="comment">## 150</span></span><br><span class="line">[15:55:16] qwzhou@login:~/practice/hepG2/WGBS :</span><br><span class="line">$ scp ./hepG2.snp*haplo* qwzhou@211.69.141.142:~/project/methhap/WGBS/hepG2</span><br><span class="line"></span><br><span class="line">[16:11:46] qwzhou@login:~/practice/Imr90_epi/encode/BSseq :</span><br><span class="line">$ scp ./imr90.snp*haplo* qwzhou@211.69.141.142:~/project/methhap/WGBS/imr90</span><br><span class="line"></span><br><span class="line">[15:57:52] qwzhou@login:~/practice/k562/WGBS :</span><br><span class="line">$ scp ./k562.snp*haplo* qwzhou@211.69.141.142:~/project/methhap/WGBS/k562</span><br><span class="line"></span><br><span class="line">[16:10:31] qwzhou@login:~/practice/A549_BS :</span><br><span class="line">$ scp ./bshichap.cpp qwzhou@211.69.141.142:~/project/methhap/WGBS/</span><br></pre></td></tr></table></figure><p>– 单倍型长度<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### imr90 2k 1k 500bp</span></span><br><span class="line"><span class="comment">## meth/snp</span></span><br><span class="line">[16:06:02] qwzhou@mn02:~/project/methhap/WGBS/imr90 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> imr90.snpwithmr.p.hapcut2.haplo.txt imr90.snpwithmr.n.hapcut2.haplo.txt  | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | sort -k1,1 -k2,2n -k3,3n | bedtools merge &gt; imr90.hapcut2.haplo.merge.bed</span><br><span class="line">[16:07:48] qwzhou@mn02:~/project/methhap/WGBS/imr90 :</span><br><span class="line">$ awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span> imr90.hapcut2.haplo.merge.bed</span><br><span class="line">1525 13531 77132</span><br><span class="line"><span class="comment">## snp</span></span><br><span class="line">[16:08:00] qwzhou@mn02:~/project/methhap/WGBS/imr90 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> imr90.snp.hapcut2.haplo.txt | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | sort -k1,1 -k2,2n -k3,3n | awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span></span><br><span class="line">22 235 3053</span><br><span class="line"><span class="comment">## meth +snp, snp</span></span><br><span class="line">[16:15:37] qwzhou@mn02:~/project/methhap/WGBS/imr90 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'ARGIND==1&#123;v[$1" "$2]&#125;ARGIND!=1 &amp;&amp; $1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0 &amp;&amp; pp==1)&#123;print "=== "chr,start,end"\n"inf&#125;;pp=0;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0 &amp;&amp; pp==1)&#123;print "=== "chr,start,end"\n"inf&#125;;pp=0;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if($4" "$5 in v)&#123;pp=1&#125;;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> wgbs.replace.mdups.snp.filtered.sort.vcf imr90.snpwithmr.p.hapcut2.haplo.txt imr90.snpwithmr.n.hapcut2.haplo.txt  | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | sort -k1,1 -k2,2n -k3,3n | bedtools merge | awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span></span><br><span class="line">1186 9720 51002</span><br><span class="line"></span><br><span class="line"><span class="comment">#### total len</span></span><br><span class="line"><span class="comment">## meth/snp</span></span><br><span class="line">[16:11:36] qwzhou@mn02:~/project/methhap/WGBS/imr90 :</span><br><span class="line">$ awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span> imr90.hapcut2.haplo.merge.bed</span><br><span class="line">278401202 1571037 177.209</span><br><span class="line"><span class="comment">## snp</span></span><br><span class="line">[16:11:59] qwzhou@mn02:~/project/methhap/WGBS/imr90 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> imr90.snp.hapcut2.haplo.txt | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | sort -k1,1 -k2,2n -k3,3n |  awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span></span><br><span class="line">30820961 227502 135.476</span><br><span class="line"><span class="comment">## snp +meth, snp</span></span><br><span class="line">[16:12:33] qwzhou@mn02:~/project/methhap/WGBS/imr90 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'ARGIND==1&#123;v[$1" "$2]&#125;ARGIND!=1 &amp;&amp; $1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0 &amp;&amp; pp==1)&#123;print "=== "chr,start,end"\n"inf&#125;;pp=0;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0 &amp;&amp; pp==1)&#123;print "=== "chr,start,end"\n"inf&#125;;pp=0;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if($4" "$5 in v)&#123;pp=1&#125;;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> wgbs.replace.mdups.snp.filtered.sort.vcf imr90.snpwithmr.p.hapcut2.haplo.txt imr90.snpwithmr.n.hapcut2.haplo.txt  | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | sort -k1,1 -k2,2n -k3,3n | bedtools merge | awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span></span><br><span class="line">149721774 681489 219.698</span><br></pre></td></tr></table></figure><p></p><p>—合并hap<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 合并meth</span></span><br><span class="line">[14:57:07] qwzhou@mn02:~/project/methhap/WGBS/imr90 :</span><br><span class="line">$ ../bsmerge imr90.snpwithmr.p.hapcut2.haplo.txt imr90.snpwithmr.n.hapcut2.haplo.txt &gt; imr90.snpwithmr.merge.hapcut2.haplo.txt</span><br><span class="line"><span class="comment">##合并hic</span></span><br><span class="line">[14:59:00] qwzhou@mn02:~/project/methhap/WGBS/imr90 :</span><br><span class="line">$  ../bsmergehic imr90.snpwithmr.merge.hapcut2.haplo.txt ../../hic/imr90/hic_out/bowtie_results/bwt2/imr90.haplo.txt &gt; hap.mergedhic.txt</span><br></pre></td></tr></table></figure><p></p><p>统计长度</p><h2 id="hic"><a href="#hic" class="headerlink" title="+ hic"></a>+ hic</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">hicmerge=<span class="string">"hap.mergedhic.txt"</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'BEGIN&#123;a=0;b=0;chr=""&#125;&#123;</span></span><br><span class="line"><span class="string">     if($1~/^B/)&#123;</span></span><br><span class="line"><span class="string">         if(b&gt;a)&#123;print chr,a,b;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">         else&#123;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">     else if(a==0)&#123;</span></span><br><span class="line"><span class="string">         chr=$1;a=$2;b=$2</span></span><br><span class="line"><span class="string">     &#125;else&#123;</span></span><br><span class="line"><span class="string">         if($2&gt;b+3000)&#123;</span></span><br><span class="line"><span class="string">             if(b&gt;a)&#123;print chr,a,b; a=$2;b=$2;chr=$1&#125;</span></span><br><span class="line"><span class="string">         &#125;else if($2&gt;b)&#123;</span></span><br><span class="line"><span class="string">             b=$2</span></span><br><span class="line"><span class="string">         &#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(b&gt;a)&#123;print chr,a,b&#125;&#125;'</span> <span class="variable">$&#123;hicmerge&#125;</span> | awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span></span><br><span class="line">331143923 1823085 181.639</span><br><span class="line"></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'BEGIN&#123;a=0;b=0;chr=""&#125;&#123;</span></span><br><span class="line"><span class="string">     if($1~/^B/)&#123;</span></span><br><span class="line"><span class="string">         if(b&gt;a)&#123;print chr,a,b;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">         else&#123;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">     else if(a==0)&#123;</span></span><br><span class="line"><span class="string">         chr=$1;a=$2;b=$2</span></span><br><span class="line"><span class="string">     &#125;else&#123;</span></span><br><span class="line"><span class="string">         if($2&gt;b+3000)&#123;</span></span><br><span class="line"><span class="string">             if(b&gt;a)&#123;print chr,a,b; a=$2;b=$2;chr=$1&#125;</span></span><br><span class="line"><span class="string">         &#125;else if($2&gt;b)&#123;</span></span><br><span class="line"><span class="string">             b=$2</span></span><br><span class="line"><span class="string">         &#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(b&gt;a)&#123;print chr,a,b&#125;&#125;'</span> <span class="variable">$&#123;hicmerge&#125;</span> | awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span></span><br><span class="line">2136 15067 91041</span><br></pre></td></tr></table></figure><h3 id="覆盖repeat情况"><a href="#覆盖repeat情况" class="headerlink" title="覆盖repeat情况"></a>覆盖repeat情况</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### + hic</span></span><br><span class="line">hicmerge=<span class="string">"A549.snpwithmr.merge.hic.haplo.txt"</span></span><br><span class="line">snpmrmerge=<span class="string">"A549.snpwithmr.haplo.merge.txt"</span></span><br><span class="line">mrmerge=<span class="string">"A549.mr.haplo.merge.txt"</span></span><br><span class="line">snphap=<span class="string">"A549.snv.hapcut2.haplo.txt"</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'BEGIN&#123;a=0;b=0;chr=""&#125;&#123;</span></span><br><span class="line"><span class="string">     if($1~/^B/)&#123;</span></span><br><span class="line"><span class="string">         if(b&gt;a)&#123;print chr,a,b;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">         else&#123;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">     else if(a==0)&#123;</span></span><br><span class="line"><span class="string">         chr=$1;a=$2;b=$2</span></span><br><span class="line"><span class="string">     &#125;else&#123;</span></span><br><span class="line"><span class="string">         if($2&gt;b+3000)&#123;</span></span><br><span class="line"><span class="string">             if(b&gt;a)&#123;print chr,a,b; a=$2;b=$2;chr=$1&#125;</span></span><br><span class="line"><span class="string">         &#125;else if($2&gt;b)&#123;</span></span><br><span class="line"><span class="string">             b=$2</span></span><br><span class="line"><span class="string">         &#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(b&gt;a)&#123;print chr,a,b&#125;&#125;'</span> <span class="variable">$&#123;hicmerge&#125;</span> | bedtools intersect -a - -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line"><span class="comment">## meth snp</span></span><br><span class="line">bedtools intersect -a <span class="variable">$&#123;snpmrmerge&#125;</span> -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line"><span class="comment">## meth</span></span><br><span class="line">bedtools intersect -a <span class="variable">$&#123;mrmerge&#125;</span> -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line"><span class="comment">## snp</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> <span class="variable">$&#123;snphap&#125;</span> | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | bedtools intersect -a - -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br></pre></td></tr></table></figure><h3 id="整合成脚本"><a href="#整合成脚本" class="headerlink" title="整合成脚本"></a>整合成脚本</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> WGBS/imr90/</span><br><span class="line">methsnpp=<span class="string">"imr90.snpwithmr.p.hapcut2.haplo.txt"</span></span><br><span class="line">methsnpn=<span class="string">"imr90.snpwithmr.n.hapcut2.haplo.txt"</span></span><br><span class="line">methp=<span class="string">"imr90.mr.hapcut2.p.haplo.txt"</span></span><br><span class="line">methn=<span class="string">"imr90.mr.hapcut2.n.haplo.txt"</span></span><br><span class="line">snphap=<span class="string">"imr90.snp.hapcut2.haplo.txt"</span></span><br><span class="line"><span class="comment">## meth + snp</span></span><br><span class="line">methsnpmerge=<span class="string">"snpwithmr.haplo.merge.txt"</span></span><br><span class="line">methmerge=<span class="string">"mr.haplo.merge.txt"</span></span><br><span class="line">hichap=<span class="string">"/public/home/qwzhou/project/methhap/hic/imr90/hic_out/bowtie_results/bwt2/imr90.haplo.txt"</span></span><br><span class="line"></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf</span></span><br><span class="line"><span class="string">&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end</span></span><br><span class="line"><span class="string">"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\</span></span><br><span class="line"><span class="string">t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> <span class="variable">$&#123;methsnpp&#125;</span> $&#123;methsn</span><br><span class="line">pp&#125; | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | sort -k1,1 -k2,2n -k3,3n | bedtools merge &gt; <span class="variable">$&#123;methsnpmerge&#125;</span></span><br><span class="line"></span><br><span class="line">awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span> <span class="variable">$&#123;methsnpmerge&#125;</span></span><br><span class="line">awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span> <span class="variable">$&#123;methsnpmerge&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## meth</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf</span></span><br><span class="line"><span class="string">&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end</span></span><br><span class="line"><span class="string">"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\</span></span><br><span class="line"><span class="string">t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> <span class="variable">$&#123;methp&#125;</span> <span class="variable">$&#123;methp&#125;</span>  |</span><br><span class="line"> awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | sort -k1,1 -k2,2n -k3,3n | bedtools merge &gt; <span class="variable">$&#123;methmerge&#125;</span></span><br><span class="line">awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span> <span class="variable">$&#123;methmerge&#125;</span></span><br><span class="line">awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span> <span class="variable">$&#123;methmerge&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## snp</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> <span class="variable">$&#123;snphap&#125;</span> | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span></span><br><span class="line"></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> <span class="variable">$&#123;snphap&#125;</span> | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## snp meth +hic</span></span><br><span class="line"><span class="comment">##合并snpmeth</span></span><br><span class="line">~/project/methhap/WGBS/bsmerge <span class="variable">$&#123;methsnpp&#125;</span> <span class="variable">$&#123;methsnpn&#125;</span> &gt; snpwithmr.merge.hapcut2.haplo.txt</span><br><span class="line"><span class="comment">##合并hic</span></span><br><span class="line">~/project/methhap/WGBS/bsmergehic snpwithmr.merge.hapcut2.haplo.txt <span class="variable">$&#123;hichap&#125;</span> &gt; snpwithmr.hap.mergedhic.txt</span><br><span class="line"></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'BEGIN&#123;a=0;b=0;chr=""&#125;&#123;</span></span><br><span class="line"><span class="string">     if($1~/^B/)&#123;</span></span><br><span class="line"><span class="string">         if(b&gt;a)&#123;print chr,a,b;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">         else&#123;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">     else if(a==0)&#123;</span></span><br><span class="line"><span class="string">         chr=$1;a=$2;b=$2</span></span><br><span class="line"><span class="string">     &#125;else&#123;</span></span><br><span class="line"><span class="string">         if($2&gt;b+3000)&#123;</span></span><br><span class="line"><span class="string">             if(b&gt;a)&#123;print chr,a,b; a=$2;b=$2;chr=$1&#125;</span></span><br><span class="line"><span class="string">         &#125;else if($2&gt;b)&#123;</span></span><br><span class="line"><span class="string">             b=$2</span></span><br><span class="line"><span class="string">         &#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(b&gt;a)&#123;print chr,a,b&#125;&#125;'</span> snpwithmr.hap.mergedhic.txt | awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span></span><br><span class="line"></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'BEGIN&#123;a=0;b=0;chr=""&#125;&#123;</span></span><br><span class="line"><span class="string">     if($1~/^B/)&#123;</span></span><br><span class="line"><span class="string">         if(b&gt;a)&#123;print chr,a,b;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">         else&#123;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">     else if(a==0)&#123;</span></span><br><span class="line"><span class="string">         chr=$1;a=$2;b=$2</span></span><br><span class="line"><span class="string">     &#125;else&#123;</span></span><br><span class="line"><span class="string">         if($2&gt;b+3000)&#123;</span></span><br><span class="line"><span class="string">             if(b&gt;a)&#123;print chr,a,b; a=$2;b=$2;chr=$1&#125;</span></span><br><span class="line"><span class="string">         &#125;else if($2&gt;b)&#123;</span></span><br><span class="line"><span class="string">             b=$2</span></span><br><span class="line"><span class="string">         &#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(b&gt;a)&#123;print chr,a,b&#125;&#125;'</span> snpwithmr.hap.mergedhic.txt | awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span></span><br><span class="line"></span><br><span class="line"><span class="comment">### 覆盖 repeat</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## meth snp</span></span><br><span class="line">bedtools intersect -a <span class="variable">$&#123;methsnpmerge&#125;</span> -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## meth</span></span><br><span class="line">bedtools intersect -a <span class="variable">$&#123;methmerge&#125;</span> -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## snp</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> <span class="variable">$&#123;snphap&#125;</span> | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | bedtools intersect -a - -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line"></span><br><span class="line"><span class="comment">### + hic</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'BEGIN&#123;a=0;b=0;chr=""&#125;&#123;</span></span><br><span class="line"><span class="string">     if($1~/^B/)&#123;</span></span><br><span class="line"><span class="string">         if(b&gt;a)&#123;print chr,a,b;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">         else&#123;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">     else if(a==0)&#123;</span></span><br><span class="line"><span class="string">         chr=$1;a=$2;b=$2</span></span><br><span class="line"><span class="string">     &#125;else&#123;</span></span><br><span class="line"><span class="string">         if($2&gt;b+3000)&#123;</span></span><br><span class="line"><span class="string">             if(b&gt;a)&#123;print chr,a,b; a=$2;b=$2;chr=$1&#125;</span></span><br><span class="line"><span class="string">         &#125;else if($2&gt;b)&#123;</span></span><br><span class="line"><span class="string">             b=$2</span></span><br><span class="line"><span class="string">         &#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(b&gt;a)&#123;print chr,a,b&#125;&#125;'</span> snpwithmr.hap.mergedhic.txt | bedtools intersect -a - -b ~/project/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br></pre></td></tr></table></figure><h3 id="ASM-on-exon-脚本部分"><a href="#ASM-on-exon-脚本部分" class="headerlink" title="ASM on exon 脚本部分"></a>ASM on exon 脚本部分</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">readfile(<span class="string">"/Users/qiangweizhou/K562.asmonExonend.0.02.Methy.1.txt"</span>, y,z, k)</span><br><span class="line">readfile(<span class="string">"/Users/qiangweizhou/IMR90.asmonExonend.0.02.Methy.1.txt"</span>, y,z, k)</span><br><span class="line">readfile(<span class="string">"/Users/qiangweizhou/HepG2.asmonExonend.0.02.Methy.1.txt"</span>, y,z, k)</span><br><span class="line"></span><br><span class="line">readfile(<span class="string">"/Users/qiangweizhou/A549.asmonExonend.0.02.Methy.1.txt"</span>, y,z, k)</span><br><span class="line">readfile(<span class="string">"/Users/qiangweizhou/HUES64.asmonExonend.0.02.Methy.1.txt"</span>, y,z, k)</span><br><span class="line">readfile(<span class="string">"/Users/qiangweizhou/GM12878.asmonExonend.0.02.Methy.1.txt"</span>, y,z, k)</span><br><span class="line">x = np.linspace(1, len(y[0]), len(y[0]))</span><br><span class="line"></span><br><span class="line">label=[<span class="string">'K562'</span>, <span class="string">'IMR90'</span>, <span class="string">'HepG2'</span>, <span class="string">'A549'</span>, <span class="string">'HUES64'</span>, <span class="string">'GM12878'</span>]</span><br><span class="line">filename=<span class="string">"ASMonExonE.all"</span></span><br></pre></td></tr></table></figure><h2 id="工作"><a href="#工作" class="headerlink" title="工作"></a>工作</h2><p>1 整理拟南芥数据结果。</p><p>拟南芥单倍型长度、覆盖repeat数目、ASM</p><p>2 HiC + BS 单倍型组装</p><p>IMR90 K562 etc. HiC haplotype</p><p>根据SNP信息，整合单倍型结果</p><p>3 K562 HepG2 DNA单倍型文章组装DNA单倍型</p><h3 id="k562-HepG2-单倍型与WGS比较"><a href="#k562-HepG2-单倍型与WGS比较" class="headerlink" title="k562 HepG2 单倍型与WGS比较"></a>k562 HepG2 单倍型与WGS比较</h3><h4 id="首先完成WGBS-meth-snv检测单倍型"><a href="#首先完成WGBS-meth-snv检测单倍型" class="headerlink" title="首先完成WGBS meth+snv检测单倍型"></a>首先完成WGBS meth+snv检测单倍型</h4><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">[10:45:12] qwzhou@login:~/practice/k562/WGBS :</span><br><span class="line"><span class="meta">$</span><span class="bash"> <span class="built_in">echo</span> <span class="string">'sh run.hapcut2.sh'</span> | qsub -l nodes=1:ppn=8 -d ./ -e ./ -N k562</span></span><br></pre></td></tr></table></figure><h4 id="WGS单倍型"><a href="#WGS单倍型" class="headerlink" title="WGS单倍型"></a>WGS单倍型</h4><p>首先将下载的文件转为hg38基因组对应。过程查看dayday-liftover部分。<br>先比较一下我们k562 vcf文件与结果中的信息重合度。<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">[14:27:14] qwzhou@login:~/practice/k562/WGBS :</span><br><span class="line">$ awk <span class="string">'ARGIND==1&#123;a[$1" "$2]&#125;ARGIND==2&#123;if($1" "$2 in a)&#123;print $0&#125;&#125;'</span> wgbs.replace.mdups.snp.filtered.sort.vcf k562.wgshap.bed |wc -l</span><br><span class="line">944306</span><br><span class="line">[14:27:47] qwzhou@login:~/practice/k562/WGBS :</span><br><span class="line">$ wc -l k562.wgshap.bed</span><br><span class="line">1178703 k562.wgshap.bed</span><br></pre></td></tr></table></figure><p></p><p>重合情况还算不错<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">'ARGIND==1&#123;a[$1" "$2]=$4;b[$1" "$2]=$5&#125;ARGIND==2&#123;</span></span><br><span class="line"><span class="string">    if($1~/^B/)&#123;if(s1&gt;s2)&#123;same+=s1;diff+=d1&#125;else&#123;same+=s2;diff+=d2&#125;; s1=0;d1=0;s2=0;d2=0;&#125;else if($2==0 || $2==1)&#123;</span></span><br><span class="line"><span class="string">        if($2==0)&#123;</span></span><br><span class="line"><span class="string">            if($4" "$5 in a)&#123;</span></span><br><span class="line"><span class="string">                if(a[$4" "$5]==$6)&#123;s1++&#125;else&#123;d1++&#125;</span></span><br><span class="line"><span class="string">                if(b[$4" "$5]==$6)&#123;s2++&#125;else&#123;d2++&#125;</span></span><br><span class="line"><span class="string">            &#125;</span></span><br><span class="line"><span class="string">        &#125;else&#123;</span></span><br><span class="line"><span class="string">            if($4" "$5 in a)&#123;</span></span><br><span class="line"><span class="string">                if(a[$4" "$5]==$7)&#123;s1++&#125;else&#123;d1++&#125; </span></span><br><span class="line"><span class="string">                if(b[$4" "$5]==$7)&#123;s2++&#125;else&#123;d2++&#125;</span></span><br><span class="line"><span class="string">            &#125; </span></span><br><span class="line"><span class="string">        &#125; </span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(s1&gt;s2)&#123;same+=s1;diff+=d1&#125;else&#123;same+=s2;diff+=d2&#125;;print same, diff&#125;'</span> k562.wgshap.bed k562.snp.hapcut2.haplo.txt </span><br><span class="line">196340 22153</span><br></pre></td></tr></table></figure><p></p><p>杂合SNP的数目：1579372<br>杂合SNP与组装后单倍型重合数目：663911<br>WGS单倍型杂合SNP数目：577474<br>WGS杂合SNP与bs snp重合数目：479644<br>WGS与单倍型重合数目：217711， 其中相同的196340，差异的22153 | 21371<br>mrsnp与wgs单倍型：344913,<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">[15:04:30] qwzhou@comput54:~/practice/k562/WGBS :</span><br><span class="line">$ awk <span class="string">'ARGIND==1 &amp;&amp; $NF~/0\/1/ &amp;&amp; $7=="PASS"'</span> wgbs.replace.mdups.snp.filtered.sort.vcf | wc -l</span><br><span class="line">1579372</span><br><span class="line">awk <span class="string">'ARGIND==1 &amp;&amp; $NF~/0\/1/ &amp;&amp; $7=="PASS"&#123;a[$1" "$2]&#125;ARGIND==2&#123;if($2==0 || $2==1)&#123;if($4" "$5 in a)&#123;print $0&#125;&#125;&#125;'</span> wgbs.replace.mdups.snp.filtered.sort.vcf k562.snp.hapcut2.haplo.txt |wc -l</span><br><span class="line">663911</span><br><span class="line">awk <span class="string">'ARGIND==1 &amp;&amp; $4!=$5'</span> k562.wgshap.bed |wc -l</span><br><span class="line">577474</span><br><span class="line">awk <span class="string">'ARGIND==1 &amp;&amp; $NF~/0\/1/ &amp;&amp; $7=="PASS"&#123;a[$1" "$2]&#125;ARGIND==2 &amp;&amp; $4!=$5&#123;if($1" "$2 in a)&#123;print $0&#125;&#125;'</span> wgbs.replace.mdups.snp.filtered.sort.vcf k562.wgshap.bed |wc -l</span><br><span class="line">479644</span><br><span class="line">awk <span class="string">'ARGIND==1 &amp;&amp; $4!=$5&#123;a[$1" "$2]&#125;ARGIND==2&#123;if($2==0 || $2==1)&#123;if($4" "$5 in a)&#123;print $0&#125;&#125;&#125;'</span> k562.wgshap.bed k562.snp.hapcut2.haplo.txt |wc -l</span><br><span class="line">217711</span><br></pre></td></tr></table></figure><p></p><p>计算methSNP重合度：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">'BEGIN&#123;split("",s1);split("",d1);split("",s2);split("",d2);&#125;</span></span><br><span class="line"><span class="string">    ARGIND==1&#123;a[$1" "$2]=$4;b[$1" "$2]=$5&#125;ARGIND!=1&#123;</span></span><br><span class="line"><span class="string">    if($1~/^B/ &amp;&amp; FNR&gt;1)&#123;</span></span><br><span class="line"><span class="string">        if(length(s1)&gt;length(s2))&#123;</span></span><br><span class="line"><span class="string">            for(i in s1)&#123;same[i]=s1[i]&#125;;for(i in d1)&#123;diff[i]=d1[i]&#125;</span></span><br><span class="line"><span class="string">        &#125;else&#123;</span></span><br><span class="line"><span class="string">            for(i in s2)&#123;same[i]=s2[i]&#125;;for(i in d2)&#123;diff[i]=d2[i]&#125;;</span></span><br><span class="line"><span class="string">        &#125;</span></span><br><span class="line"><span class="string">        split("",s1);split("",d1);split("",s2);split("",d2);</span></span><br><span class="line"><span class="string">    &#125;else if($2==0 || $2==1)&#123;</span></span><br><span class="line"><span class="string">        if($2==0)&#123;</span></span><br><span class="line"><span class="string">            if($4" "$5 in a)&#123;</span></span><br><span class="line"><span class="string">                if(a[$4" "$5]==$6)&#123;s1[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d1[$4" "$5]&#125;&#125;</span></span><br><span class="line"><span class="string">                if(b[$4" "$5]==$6)&#123;s2[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d2[$4" "$5]&#125;&#125;</span></span><br><span class="line"><span class="string">            &#125;</span></span><br><span class="line"><span class="string">        &#125;else&#123;</span></span><br><span class="line"><span class="string">            if($4" "$5 in a)&#123;</span></span><br><span class="line"><span class="string">                if(a[$4" "$5]==$7)&#123;s1[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d1[$4" "$5]&#125;&#125; </span></span><br><span class="line"><span class="string">                if(b[$4" "$5]==$7)&#123;s2[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d2[$4" "$5]&#125;&#125;</span></span><br><span class="line"><span class="string">            &#125; </span></span><br><span class="line"><span class="string">        &#125; </span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;</span></span><br><span class="line"><span class="string">    if(length(s1)&gt;length(s2))&#123;</span></span><br><span class="line"><span class="string">        for(i in s1)&#123;same[i]=s1[i]&#125;;for(i in d1)&#123;diff[i]=d1[i]&#125;</span></span><br><span class="line"><span class="string">    &#125;else&#123;</span></span><br><span class="line"><span class="string">        for(i in s2)&#123;same[i]=s2[i]&#125;;for(i in d2)&#123;diff[i]=d2[i]&#125;;</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">    print length(same), length(diff)</span></span><br><span class="line"><span class="string">&#125;'</span> k562.wgshap.bed k562.snpwithmr.p.hapcut2.haplo.txt k562.snpwithmr.n.hapcut2.haplo.txt </span><br><span class="line">335646 36349</span><br></pre></td></tr></table></figure><p></p><p>统计mapQ==100的hap：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">'BEGIN&#123;split("",s1);split("",d1);split("",s2);split("",d2);&#125;</span></span><br><span class="line"><span class="string">    ARGIND==1&#123;a[$1" "$2]=$4;b[$1" "$2]=$5&#125;ARGIND!=1&#123;</span></span><br><span class="line"><span class="string">    if($1~/^B/ &amp;&amp; FNR&gt;1)&#123;</span></span><br><span class="line"><span class="string">        if(length(s1)&gt;length(s2))&#123;</span></span><br><span class="line"><span class="string">            for(i in s1)&#123;same[i]=s1[i]&#125;;for(i in d1)&#123;diff[i]=d1[i]&#125;</span></span><br><span class="line"><span class="string">        &#125;else&#123;</span></span><br><span class="line"><span class="string">            for(i in s2)&#123;same[i]=s2[i]&#125;;for(i in d2)&#123;diff[i]=d2[i]&#125;;</span></span><br><span class="line"><span class="string">        &#125;</span></span><br><span class="line"><span class="string">        split("",s1);split("",d1);split("",s2);split("",d2);</span></span><br><span class="line"><span class="string">    &#125;else if($NF==100 &amp;&amp; ($2==0 || $2==1))&#123;</span></span><br><span class="line"><span class="string">        if($2==0)&#123;</span></span><br><span class="line"><span class="string">            if($4" "$5 in a)&#123;</span></span><br><span class="line"><span class="string">                if(a[$4" "$5]==$6)&#123;s1[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d1[$4" "$5]&#125;&#125;</span></span><br><span class="line"><span class="string">                if(b[$4" "$5]==$6)&#123;s2[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d2[$4" "$5]&#125;&#125;</span></span><br><span class="line"><span class="string">            &#125;</span></span><br><span class="line"><span class="string">        &#125;else&#123;</span></span><br><span class="line"><span class="string">            if($4" "$5 in a)&#123;</span></span><br><span class="line"><span class="string">                if(a[$4" "$5]==$7)&#123;s1[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d1[$4" "$5]&#125;&#125;</span></span><br><span class="line"><span class="string">                if(b[$4" "$5]==$7)&#123;s2[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d2[$4" "$5]&#125;&#125;</span></span><br><span class="line"><span class="string">            &#125;</span></span><br><span class="line"><span class="string">        &#125;</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;</span></span><br><span class="line"><span class="string">    if(length(s1)&gt;length(s2))&#123;</span></span><br><span class="line"><span class="string">        for(i in s1)&#123;same[i]=s1[i]&#125;;for(i in d1)&#123;diff[i]=d1[i]&#125;</span></span><br><span class="line"><span class="string">    &#125;else&#123;</span></span><br><span class="line"><span class="string">        for(i in s2)&#123;same[i]=s2[i]&#125;;for(i in d2)&#123;diff[i]=d2[i]&#125;;</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">    print length(same), length(diff)</span></span><br><span class="line"><span class="string">&#125;'</span> k562.wgshap.bed k562.snpwithmr.p.hapcut2.haplo.txt k562.snpwithmr.n.hapcut2.haplo.txt</span><br></pre></td></tr></table></figure><p></p><h4 id="统计k562、hepG2与WGS-hap重合位点，方便可视化。"><a href="#统计k562、hepG2与WGS-hap重合位点，方便可视化。" class="headerlink" title="统计k562、hepG2与WGS hap重合位点，方便可视化。"></a>统计k562、hepG2与WGS hap重合位点，方便可视化。</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">'BEGIN&#123;split("",s1);split("",d1);split("",s2);split("",d2);&#125;</span></span><br><span class="line"><span class="string">    ARGIND==1&#123;a[$1" "$2]=$4;b[$1" "$2]=$5&#125;ARGIND!=1&#123;</span></span><br><span class="line"><span class="string">    if($1~/^B/ &amp;&amp; FNR&gt;1)&#123;</span></span><br><span class="line"><span class="string">        if(length(s1)&gt;length(s2))&#123;</span></span><br><span class="line"><span class="string">            for(i in s1)&#123;same[i]=s1[i]&#125;;for(i in d1)&#123;diff[i]=d1[i]&#125;</span></span><br><span class="line"><span class="string">        &#125;else&#123;</span></span><br><span class="line"><span class="string">            for(i in s2)&#123;same[i]=s2[i]&#125;;for(i in d2)&#123;diff[i]=d2[i]&#125;;</span></span><br><span class="line"><span class="string">        &#125;</span></span><br><span class="line"><span class="string">        split("",s1);split("",d1);split("",s2);split("",d2);</span></span><br><span class="line"><span class="string">    &#125;else if($2==0 || $2==1)&#123;</span></span><br><span class="line"><span class="string">        if($2==0)&#123;</span></span><br><span class="line"><span class="string">            if($4" "$5 in a)&#123;</span></span><br><span class="line"><span class="string">                if(a[$4" "$5]==$6)&#123;s1[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d1[$4" "$5]&#125;&#125;</span></span><br><span class="line"><span class="string">                if(b[$4" "$5]==$6)&#123;s2[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d2[$4" "$5]&#125;&#125;</span></span><br><span class="line"><span class="string">            &#125;</span></span><br><span class="line"><span class="string">        &#125;else&#123;</span></span><br><span class="line"><span class="string">            if($4" "$5 in a)&#123;</span></span><br><span class="line"><span class="string">                if(a[$4" "$5]==$7)&#123;s1[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d1[$4" "$5]&#125;&#125;</span></span><br><span class="line"><span class="string">                if(b[$4" "$5]==$7)&#123;s2[$4" "$5]&#125;else&#123;if($8~/:/)&#123;d2[$4" "$5]&#125;&#125;</span></span><br><span class="line"><span class="string">            &#125;</span></span><br><span class="line"><span class="string">        &#125;</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;</span></span><br><span class="line"><span class="string">    if(length(s1)&gt;length(s2))&#123;</span></span><br><span class="line"><span class="string">        for(i in s1)&#123;same[i]=s1[i]&#125;;for(i in d1)&#123;diff[i]=d1[i]&#125;</span></span><br><span class="line"><span class="string">    &#125;else&#123;</span></span><br><span class="line"><span class="string">        for(i in s2)&#123;same[i]=s2[i]&#125;;for(i in d2)&#123;diff[i]=d2[i]&#125;;</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">    for(i in same)&#123;print i,"same"&#125;;for(i in diff)&#123;print i,"diff"&#125;</span></span><br><span class="line"><span class="string">&#125;'</span> k562.wgshap.bed k562.snp.hapcut2.haplo.txt | sort -k1,1 -k2,2n &gt; k562.snphap.wgshap.txt</span><br></pre></td></tr></table></figure><p>snpmr:<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">...</span><br><span class="line">k562.wgshap.bed k562.snpwithmr.p.hapcut2.haplo.txt k562.snpwithmr.n.hapcut2.haplo.txt | sort -k1,1 -k2,2n &gt; k562.mrsnphap.wgshap.txt</span><br></pre></td></tr></table></figure><p></p><p>合并文件：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">[15:14:49] qwzhou@login:~/practice/k562/WGBS :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'ARGIND==1&#123;a[$1"\t"$2]="N\tN"&#125;ARGIND==2&#123;a[$1"\t"$2]=$3"\tN"&#125;ARGIND==3&#123;split(a[$1"\t"$2],b,"\t"); a[$1"\t"$2]=b[1]"\t"$3&#125;END&#123;for(i in a)&#123;print i,a[i]&#125;&#125;'</span> k562.wgshap.bed k562.mrsnphap.wgshap.txt k562.snphap.wgshap.txt |sort -k1,1 -k2,2n &gt; k562.vs.wgshap.merge.txt</span><br></pre></td></tr></table></figure><p></p><p>数据太大没法可视化，做进一步处理：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">╭─[MacBook-Pro] as qiangweizhou <span class="keyword">in</span> ~/科研/DNA甲基化单倍型/bshichaplo                                 04-11 16:19:43</span><br><span class="line">╰──➤  awk -v OFS=<span class="string">"\t"</span> <span class="string">'$3!="N" &amp;&amp; $4!="N"&#123;a[$1"\t"int($2/2000)"\t"$3]++; b[$1"\t"int($2/2000)"\t"$4]++; id[$1"\t"int($2/2000)];&#125;</span></span><br><span class="line"><span class="string">END&#123;</span></span><br><span class="line"><span class="string">    for(i in id)&#123;</span></span><br><span class="line"><span class="string">        if(!(i"\tsame" in a))&#123;a[i"\tsame"]=0&#125;</span></span><br><span class="line"><span class="string">        if(!(i"\tdiff" in a))&#123;a[i"\tdiff"]=0&#125;</span></span><br><span class="line"><span class="string">        if(!(i"\tsame" in b))&#123;b[i"\tsame"]=0&#125;</span></span><br><span class="line"><span class="string">        if(!(i"\tdiff" in b))&#123;b[i"\tdiff"]=0&#125;</span></span><br><span class="line"><span class="string">        print i,a[i"\tsame"],a[i"\tdiff"], b[i"\tsame"], b[i"\tdiff"]</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">&#125;'</span> k562.vs.wgshap.merge.txt | sort -k1,1 -k2,2n &gt; k562.vs.wgshap.merge.2kb.txt</span><br></pre></td></tr></table></figure><p></p><p>加入位置信息：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'!($3=="N" &amp;&amp; $4=="N")&#123;a[$1"\t"int($2/2000)"\t"$3]++; b[$1"\t"int($2/2000)"\t"$4]++; id[$1"\t"int($2/2000)];&#125;</span></span><br><span class="line"><span class="string">END&#123;</span></span><br><span class="line"><span class="string">    for(i in id)&#123;</span></span><br><span class="line"><span class="string">        if(!(i"\tsame" in a))&#123;a[i"\tsame"]=0&#125;</span></span><br><span class="line"><span class="string">        if(!(i"\tdiff" in a))&#123;a[i"\tdiff"]=0&#125;</span></span><br><span class="line"><span class="string">        if(!(i"\tsame" in b))&#123;b[i"\tsame"]=0&#125;</span></span><br><span class="line"><span class="string">        if(!(i"\tdiff" in b))&#123;b[i"\tdiff"]=0&#125;</span></span><br><span class="line"><span class="string">        split(i,pos,"\t");print pos[2]*2000+1,i,a[i"\tsame"],a[i"\tdiff"], b[i"\tsame"], b[i"\tdiff"]</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">&#125;'</span> k562.vs.wgshap.merge.txt | sort -k2,2 -k1,1n &gt; k562.vs.wgshap.merge.2kb.txt</span><br></pre></td></tr></table></figure><p></p><h5 id="hepG2"><a href="#hepG2" class="headerlink" title="hepG2"></a>hepG2</h5><p>格式转换<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;print "chr"$1,$2,$2+1,$3,$4&#125;'</span> HepG2.haplotyped_across_tumor.passing.clean.txt &gt; HepG2.haplotyped.clean.bed</span><br></pre></td></tr></table></figure><p></p><p>运行liftover：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">[16:19:37] qwzhou@mn02:~/hapvswgs :</span><br><span class="line">$ ../liftOver HepG2.haplotyped.clean.bed hg19ToHg38.over.chain.gz HepG2.hapwgs.bed hepG2.unlifted.bed</span><br></pre></td></tr></table></figure><p></p><p>计算重合度SNP: 72551 7711<br>计算重合度MethSNP: 125816 12258<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##脚本与k562过程一致，不再重复记录，只记录文件名这里。</span></span><br><span class="line">HepG2.hapwgs.bed hepG2.snp.hapcut2.haplo.txt &gt; hepG2.snphap.wgshap.txt</span><br><span class="line">HepG2.hapwgs.bed hepG2.snpwithmr.p.hapcut2.haplo.txt hepG2.snpwithmr.n.hapcut2.haplo.txt &gt; hepG2.mrsnphap.wgshap.txt</span><br></pre></td></tr></table></figure><p></p><p>合并文件：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">[15:14:49] qwzhou@login:~/practice/hepG2/WGBS :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'ARGIND==1&#123;a[$1"\t"$2]="N\tN"&#125;ARGIND==2&#123;a[$1"\t"$2]=$3"\tN"&#125;ARGIND==3&#123;split(a[$1"\t"$2],b,"\t"); a[$1"\t"$2]=b[1]"\t"$3&#125;END&#123;for(i in a)&#123;print i,a[i]&#125;&#125;'</span> HepG2.hapwgs.bed hepG2.mrsnphap.wgshap.txt hepG2.snphap.wgshap.txt |sort -k1,1 -k2,2n &gt; hepG2.vs.wgshap.merge.txt</span><br></pre></td></tr></table></figure><p></p><p>数据太大没法可视化，做进一步处理,加入位置信息：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'!($3=="N" &amp;&amp; $4=="N")&#123;a[$1"\t"int($2/2000)"\t"$3]++; b[$1"\t"int($2/2000)"\t"$4]++; id[$1"\t"int($2/2000)];&#125;</span></span><br><span class="line"><span class="string">END&#123;</span></span><br><span class="line"><span class="string">    for(i in id)&#123;</span></span><br><span class="line"><span class="string">        if(!(i"\tsame" in a))&#123;a[i"\tsame"]=0&#125;</span></span><br><span class="line"><span class="string">        if(!(i"\tdiff" in a))&#123;a[i"\tdiff"]=0&#125;</span></span><br><span class="line"><span class="string">        if(!(i"\tsame" in b))&#123;b[i"\tsame"]=0&#125;</span></span><br><span class="line"><span class="string">        if(!(i"\tdiff" in b))&#123;b[i"\tdiff"]=0&#125;</span></span><br><span class="line"><span class="string">        split(i,pos,"\t");print pos[2]*2000+1,i,a[i"\tsame"],a[i"\tdiff"], b[i"\tsame"], b[i"\tdiff"]</span></span><br><span class="line"><span class="string">    &#125;</span></span><br><span class="line"><span class="string">&#125;'</span> hepG2.vs.wgshap.merge.txt | sort -k2,2 -k1,1n &gt; hepG2.vs.wgshap.merge.2kb.txt</span><br></pre></td></tr></table></figure><p></p><p>尝试把间隔放大：<br>awk -v OFS=”\t” ‘!($3==”N” &amp;&amp; $4==”N”){a[$1”\t”int($2/5000)”\t”$3]++; b[$1”\t”int($2/5000)”\t”$4]++; id[$1”\t”int($2/5000)];}<br>END{<br> for(i in id){<br> if(!(i”\tsame” in a)){a[i”\tsame”]=0}<br> xif(!(i”\tdiff” in a)){a[i”\tdiff”]=0}<br> if(!(i”\tsame” in b)){b[i”\tsame”]=0}<br> if(!(i”\tdiff” in b)){b[i”\tdiff”]=0}<br> split(i,pos,”\t”);print pos[2]*5000+1,i,a[i”\tsame”],a[i”\tdiff”], b[i”\tsame”], b[i”\tdiff”]<br> }<br>}’ hepG2.vs.wgshap.merge.txt | sort -k2,2 -k1,1n &gt; hepG2.vs.wgshap.merge.5kb.txt<br>因为DNA甲基化单倍型组装是区分正负链的，因此我们去掉重复的位置来统计与WGS hap的重合情况：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">'&#123;print $1,$2,$3,$4&#125;'</span> hepG2.mrsnphap.wgshap.txt | awk <span class="string">'&#123;if($3=="same")&#123;a[$1" "$2]&#125;else if(!($1" "$2 in a))&#123;b[$1" "$2]&#125;&#125;END&#123;print length(a),length(b)&#125;'</span></span><br><span class="line">125816 3907</span><br></pre></td></tr></table></figure><p></p><p>wgs hap snp: 370768，其中杂合SNP：182255<br>snp in bs-seq: 1942278<br>snp in bs with wgs hap: 151546<br>snp单倍型with wgs单倍型: 80262, 其中相同以及差异72551 7711<br>mrsnp单倍型with wgs单倍型: 129723，相同差异：125816 3907</p><p>========================================</p><h3 id="拟南芥"><a href="#拟南芥" class="headerlink" title="拟南芥"></a>拟南芥</h3><p>目录：[10:25:19] qwzhou@login:~/practice/tair_zwang/PM : [150]<br>合并正负链，并输出到bed格式。<br></p><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS="\t" '$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;' BS-cviler.p.hapcut2.haplo.txt BS-cviler.n.hapcut2.haplo.txt  | awk -v OFS="\t" '$1~/^=/&#123;print $2,$3,$4&#125;' | bedtools merge | less</span><br></pre></td></tr></table></figure><p></p><p>替换染色体，保持一致，并去掉子类，防止重复。<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'sub(/Chr/,"",$1) &amp;&amp; $4!~/transposon_fragment/'</span> TAIR10_Transposable_Elements.bed &gt; TAIR10_Transposable_Elements.replacechr.bed</span><br></pre></td></tr></table></figure><p></p><p>bedtools merge/sort error:<br>Error: Sorted input specified, but the file - has the following out of order record</p><p>统计覆盖repeat的数目：</p><ol><li><p>DNA甲基化+SNP组装</p><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">[10:56:25] qwzhou@login:~/practice/tair_zwang/PM :</span><br><span class="line"><span class="meta">$</span><span class="bash"> awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;'</span> BS-cviler.p.hapcut2.haplo.txt BS-cviler.n.hapcut2.haplo.txt  | awk -v OFS=<span class="string">"\t"</span> <span class="string">'$1~/^=/&#123;print $2,$3,$4&#125;'</span> | sort -k1,1 -k2,2n -k3,3n | bedtools merge | bedtools intersect -a - -b TAIR10_Transposable_Elements.replacechr.bed -wo | awk <span class="string">'&#123;a[$7]&#125;END&#123;print length(a)&#125;'</span></span></span><br><span class="line">10927</span><br></pre></td></tr></table></figure></li><li><p>SNP组装</p><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS="\t" '$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;' BS-cviler.hapcut2.snp.haplo.txt  | awk -v OFS="\t" '$1~/^=/&#123;print $2,$3,$4&#125;' | sort -k1,1 -k2,2n -k3,3n | bedtools intersect -a - -b TAIR10_Transposable_Elements.replacechr.bed -wo | awk '&#123;a[$7]&#125;END&#123;print length(a)&#125;'</span><br><span class="line">3890</span><br></pre></td></tr></table></figure></li><li><p>DNA甲基化+SNP组装 (但是包含SNP)</p><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS="\t" '$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;' BS-cviler.p.hapcut2.haplo.txt BS-cviler.n.hapcut2.haplo.txt  | awk -v OFS="\t" '$1~/^=/&#123;print $2,$3,$4&#125;' | sort -k1,1 -k2,2n -k3,3n | bedtools merge | bedtools intersect -a - -b BS-cviler.snp.hetero.vcf -wo | awk -v OFS="\t" '&#123;print $1,$2,$3&#125;' |bedtools intersect -a - -b TAIR10_Transposable_Elements.replacechr.bed -wo | awk '&#123;a[$7]&#125;END&#123;print length(a)&#125;'</span><br><span class="line">6544</span><br></pre></td></tr></table></figure></li></ol><p>在基因组内的分布<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">[10:32:08] qwzhou@login:~/practice/tair_zwang/PM :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'gsub(/Chr/,"",$1) &amp;&amp; $3!="chromosome"'</span> ~/practice/Genome/arabidopsis/TAIR10_GFF3_genes.gff &gt; ~/practice/Genome/arabidopsis/TAIR10_GFF3_genes.replacechr.gff</span><br><span class="line"></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;if($7=="+")&#123;if($4&gt;2000)&#123;print $1,$4-2000,$4&#125;else&#123;print $1,1,$4&#125;&#125;else if($7=="-")&#123;print $1,$5,$5+2000&#125; &#125;'</span> ~/practice/Genome/arabidopsis/TAIR10.gene.modify.gff &gt; ~/practice/Genome/arabidopsis/TAIR10.gene.modify.2Kpromoter.bed</span><br><span class="line"></span><br><span class="line">cat BS-cviler.asm.plus.bed BS-cviler.asm.neg.bed | sort -k1,1 -k2,2n |bedtools merge | bedtools intersect -a - -b ~/practice/Genome/arabidopsis/TAIR10_GFF3_genes.replacechr.gff -f 0.50 -wo | awk <span class="string">'&#123;a[$6" "$7" "$8]&#125;END&#123;for(i in a)&#123;print i&#125;&#125;'</span> | awk <span class="string">'&#123;a[$1]++&#125;END&#123;for(i in a)&#123;print i,a[i]&#125;&#125;'</span></span><br><span class="line">ncRNA 54</span><br><span class="line">pseudogene 69</span><br><span class="line">exon 4174</span><br><span class="line">rRNA 1</span><br><span class="line">protein 3195</span><br><span class="line">tRNA 3</span><br><span class="line">transposable_element_gene 898</span><br><span class="line">mRNA_TE_gene 900</span><br><span class="line">five_prime_UTR 75</span><br><span class="line">three_prime_UTR 133</span><br><span class="line">pseudogenic_transcript 69</span><br><span class="line">mRNA 3776</span><br><span class="line">pseudogenic_exon 61</span><br><span class="line">CDS 2932</span><br><span class="line">gene 3041</span><br><span class="line">Promoter 1890</span><br><span class="line">TE 1588</span><br><span class="line">cat BS-cviler.asm.plus.bed BS-cviler.asm.neg.bed | sort -k1,1 -k2,2n |bedtools merge | bedtools intersect -a - -b ~/practice/Genome/arabidopsis/TAIR10.gene.modify.2Kpromoter.bed -f 0.50 -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line">1890</span><br><span class="line">cat BS-cviler.asm.plus.bed BS-cviler.asm.neg.bed | sort -k1,1 -k2,2n |bedtools merge | bedtools intersect -a - -b ~/practice/Genome/arabidopsis/TAIR10_Transposable_Elements.replacechr.bed -f 0.50 -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line">1588</span><br></pre></td></tr></table></figure><p></p><p>ASM在基因上的分布<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">~/software_devp/methyhaplo/ASManno -o ASMonGene -G ~/practice/Genome/arabidopsis/arabidopsis_batmeth2_index/TAIR10_chr_all.fa -gff ~/practice/Genome/arabidopsis/TAIR10.protein_coding_gene.gff -ap BS-cviler.asm.plus.bed -an BS-cviler.asm.neg.bed</span><br><span class="line"> [23:11:53] qwzhou@comput54:~/practice/tair_zwang/PM :</span><br><span class="line">$ ~/practice/A549_BS/WGBS-HAIR/ASManno -o ASMonGene -G ~/practice/Genome/arabidopsis/arabidopsis_batmeth2_index/TAIR10_chr_all.fa -gff ~/practice/Genome/arabidopsis/TAIR10.protein_coding_gene.gff -ap BS-cviler.asm.plus.bed -an BS-cviler.asm.neg.bed -s 0.01</span><br></pre></td></tr></table></figure><p></p><p>ASM基因GO分析：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">[15:34:17] qwzhou@comput54:~/practice/tair_zwang/PM :</span><br><span class="line">$ cat BS-cviler.asm.plus.bed BS-cviler.asm.neg.bed | sort -k1,1 -k2,2n -k3,3n | bedtools merge | bedtools intersect -a - -b ~/practice/Genome/arabidopsis/TAIR10.gene.modify.gff -wo | awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;print $4,$7,$8,$10,$12&#125;'</span> | awk <span class="string">'split($NF,a,";") &amp;&amp; sub(/ID=/,"",a[1])&#123;print a[1]&#125;'</span> &gt; BS-cviler.asm.gene.txt</span><br></pre></td></tr></table></figure><p></p><h4 id="A549-HiC-SNPsplit"><a href="#A549-HiC-SNPsplit" class="headerlink" title="A549 HiC SNPsplit"></a>A549 HiC SNPsplit</h4><p>根据单倍型结果区分<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">[16:30:54] qwzhou@comput54:~/practice/A549_BS/HiC_normal/hicsnp_latest_allele_withencodesnp/bowtie_results/bwt2 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'($2==0 || $2==1)&#123;if($2==0)&#123;snp=$6"/"$7&#125;else if($2==1)&#123;snp=$7"/"$6&#125;print ".",$4,$5,"1",snp&#125;'</span> A549.hic.haplotype &gt; A549.hic.haplotype.SNPsplit.vcf</span><br><span class="line"></span><br><span class="line"><span class="comment">## SNPsplit</span></span><br><span class="line">[16:31:16] qwzhou@comput54:~/practice/A549_BS/HiC_normal/hicsnp_latest_allele_withencodesnp/bowtie_results/bwt2 :</span><br><span class="line">$ ~/software/SNPsplit/SNPsplit --hic --snp_file A549.hic.haplotype.SNPsplit.vcf chr2.89771503-89802405.sortbyname.fix.bam</span><br></pre></td></tr></table></figure><p></p><h4 id="A549-ASM分布"><a href="#A549-ASM分布" class="headerlink" title="A549 ASM分布"></a>A549 ASM分布</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">[22:19:57] qwzhou@login:~/practice/Genome/hg38 :</span><br><span class="line">$ awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;if($7=="-")&#123;print $1,$5,"-"&#125;else&#123;print $1,$4,"+"&#125;&#125;'</span> gene.gtf &gt; gene.TSS.ped3</span><br><span class="line"></span><br><span class="line">[22:16:46] qwzhou@comput54:~/practice/A549_BS/WGBS-HAIR :</span><br><span class="line">$ ~/software_devp/methyhaplo/ASMannoSites -o ASMonGeneTSS -G ~/practice/Genome/hg38/batmeth2-chr/hg38.chr.fa --ped ~/practice/Genome/hg38/gene.ped3 -ap A549.newasm.plus.bed -an A549.newasm.neg.bed -s 0.01</span><br></pre></td></tr></table></figure><h4 id="统计单倍型长度"><a href="#统计单倍型长度" class="headerlink" title="统计单倍型长度"></a>统计单倍型长度</h4><ol><li><p>DNA甲基化 + SNP</p><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS="\t" '$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;' BS-cviler.p.hapcut2.haplo.txt BS-cviler.n.hapcut2.haplo.txt  | awk -v OFS="\t" '$1~/^=/&#123;print $2,$3,$4&#125;' | sort -k1,1 -k2,2n -k3,3n | bedtools merge &gt; BS-cviler.hapcut2.haplo.merge.bed</span><br><span class="line">[11:08:48] qwzhou@login:~/practice/tair_zwang/PM :</span><br><span class="line"><span class="meta">$</span><span class="bash"> awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span> BS-cviler.hapcut2.haplo.merge.bed</span></span><br><span class="line">558 1533 5535</span><br></pre></td></tr></table></figure></li><li><p>SNP</p><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS="\t" '$1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0)&#123;print "=== "chr,start,end"\n"inf&#125;;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;' BS-cviler.hapcut2.snp.haplo.txt | awk -v OFS="\t" '$1~/^=/&#123;print $2,$3,$4&#125;' | sort -k1,1 -k2,2n -k3,3n | awk '&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span><br><span class="line">14 162 785</span><br></pre></td></tr></table></figure></li><li><p>DNA甲基化+SNP组装 (但是包含SNP)</p><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS="\t" 'ARGIND==1&#123;v[$1" "$2]&#125;ARGIND!=1 &amp;&amp; $1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0 &amp;&amp; pp==1)&#123;print "=== "chr,start,end"\n"inf&#125;;pp=0;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0 &amp;&amp; pp==1)&#123;print "=== "chr,start,end"\n"inf&#125;;pp=0;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if($4" "$5 in v)&#123;pp=1&#125;;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;' BS-cviler.snp.hetero.vcf BS-cviler.p.hapcut2.haplo.txt BS-cviler.n.hapcut2.haplo.txt  | awk -v OFS="\t" '$1~/^=/&#123;print $2,$3,$4&#125;' | sort -k1,1 -k2,2n -k3,3n | bedtools merge | awk '&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span><br><span class="line">474 1325 4406</span><br></pre></td></tr></table></figure></li></ol><h4 id="统计单倍型总长度"><a href="#统计单倍型总长度" class="headerlink" title="统计单倍型总长度"></a>统计单倍型总长度</h4><ol start="2"><li>DNA甲基化+SNP (包含SNP)<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS="\t" 'ARGIND==1&#123;v[$1" "$2]&#125;ARGIND!=1 &amp;&amp; $1!~/\*/&#123;if($1~/BLOCK/)&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0 &amp;&amp; pp==1)&#123;print "=== "chr,start,end"\n"inf&#125;;pp=0;chr="";start=0;end=0;inf="";&#125;else&#123;if($2=="-")&#123;if(chr!="" &amp;&amp; end&gt;start &amp;&amp; start!=0 &amp;&amp; pp==1)&#123;print "=== "chr,start,end"\n"inf&#125;;pp=0;chr="";start=0;end=0;inf=""&#125;else if($2==0 || $2==1)&#123;if($4" "$5 in v)&#123;pp=1&#125;;if(start==0)&#123;chr=$4;start=$5;inf=$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7&#125;else&#123;inf=inf"\n"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7;if($5&gt;end)&#123;end=$5&#125;&#125;&#125;&#125;&#125;' BS-cviler.snp.hetero.vcf BS-cviler.p.hapcut2.haplo.txt BS-cviler.n.hapcut2.haplo.txt  | awk -v OFS="\t" '$1~/^=/&#123;print $2,$3,$4&#125;' | sort -k1,1 -k2,2n -k3,3n | bedtools merge | awk '&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span><br><span class="line">13787280 54362 253.62</span><br></pre></td></tr></table></figure></li></ol><h4 id="统计覆盖在基因组上phased信息"><a href="#统计覆盖在基因组上phased信息" class="headerlink" title="统计覆盖在基因组上phased信息"></a>统计覆盖在基因组上phased信息</h4><p>首先切割基因组：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;bin=2000;l=1;len=int($2/bin);for(i=0;i&lt;len;i++)&#123;print $1,i*bin+1,(i+1)*bin,l;l++&#125;;if(len*bin &lt; $2)&#123;print $1,len*bin+1, $2,l&#125;&#125;'</span> ~/practice/Genome/arabidopsis/TAIR10_chr_all.fa.fai &gt; TAIR10.splitby2k.bed</span><br></pre></td></tr></table></figure><p></p><p>处理meth haplo，因为正负链不在一个文件<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">'$1!~/^B/ &amp;&amp; $1!~/^*/&#123;print $4"\t"$5&#125;'</span> BS-cviler.p.hapcut2.haplo.txt BS-cviler.n.hapcut2.haplo.txt | sort -k1,1 -k2,2n &gt; BS-cviler.hapcut2.methsnv.sites.txt</span><br></pre></td></tr></table></figure><p></p><p>统计phased数目<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'ARGIND==1 &amp;&amp; $1!~/^#/&#123;sn[$1"\t"$2]&#125;ARGIND==2&#123;a[$4"\t"int($5/2000+1)]++&#125;ARGIND==3&#123;if($1"\t"$2 in sn)&#123;b[$1"\t"int($2/2000+1)]++&#125;&#125;ARGIND==4&#123;c[$1"\t"int($2/2000+1)]++&#125;ARGIND==5&#123;if($1"\t"$4 in a)&#123;v=a[$1"\t"$4]&#125;else&#123;v=0&#125;; if($1"\t"$4 in b)&#123;mv=b[$1"\t"$4]&#125;else&#123;mv=0&#125;;if($1"\t"$4 in c)&#123;m=c[$1"\t"$4]&#125;else&#123;m=0&#125;; print $1,$4,v,mv,m&#125;'</span> BS-cviler.snp.hetero.vcf  BS-cviler.hapcut2.snp.haplo.txt BS-cviler.hapcut2.methsnv.sites.txt BS-cviler.hapcut2.methsnv.sites.txt TAIR10.splitby2k.bed &gt; phased.txt</span><br></pre></td></tr></table></figure><p></p><p>–&gt; 2Kb太短了，换成10Kb<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## split genome</span></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;bin=10000;l=1;len=int($2/bin);for(i=0;i&lt;len;i++)&#123;print $1,i*bin+1,(i+1)*bin,l;l++&#125;;if(len*bin &lt; $2)&#123;print $1,len*bin+1, $2,l&#125;&#125;'</span> ~/practice/Genome/arabidopsis/TAIR10_chr_all.fa.fai &gt; TAIR10.splitby10k.bed</span><br><span class="line"><span class="comment">## 统计数目</span></span><br><span class="line">awk -v bin=10000 -v OFS=<span class="string">"\t"</span> <span class="string">'ARGIND==1 &amp;&amp; $1!~/^#/&#123;sn[$1"\t"$2]&#125;ARGIND==2&#123;a[$4"\t"int($5/bin+1)]++&#125;ARGIND==3&#123;if($1"\t"$2 in sn)&#123;b[$1"\t"int($2/bin+1)]++&#125;&#125;ARGIND==4&#123;c[$1"\t"int($2/bin+1)]++&#125;ARGIND==5&#123;if($1"\t"$4 in a)&#123;v=a[$1"\t"$4]&#125;else&#123;v=0&#125;; if($1"\t"$4 in b)&#123;mv=b[$1"\t"$4]&#125;else&#123;mv=0&#125;;if($1"\t"$4 in c)&#123;m=c[$1"\t"$4]&#125;else&#123;m=0&#125;; print $1,$4,v,mv,m&#125;'</span> BS-cviler.snp.hetero.vcf  BS-cviler.hapcut2.snp.haplo.txt BS-cviler.hapcut2.methsnv.sites.txt BS-cviler.hapcut2.methsnv.sites.txt TAIR10.splitby10k.bed &gt; phased.10k.txt</span><br></pre></td></tr></table></figure><p></p><p>画线图<br></p><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">a&lt;-read.table(&quot;phased.10k.txt&quot;)</span><br><span class="line">names(a)=c(&quot;chr&quot;,&quot;coord&quot;,&quot;SNV&quot;, &quot;MethSNV.SNV&quot;, &quot;MethSNV&quot;)</span><br><span class="line">a&lt;-a[a$chr==&quot;1&quot;,]</span><br><span class="line">profile_text&lt;-a[c(3,4,5)]</span><br><span class="line">library(ggplot2)</span><br><span class="line">library(reshape2)</span><br><span class="line">profile_text$xvariable = rownames(profile_text)</span><br><span class="line">data_m &lt;- melt(profile_text, id.vars=c(&quot;xvariable&quot;))</span><br><span class="line">data_m$xvariable &lt;- as.numeric(data_m$xvariable)</span><br><span class="line"></span><br><span class="line">##线图</span><br><span class="line">data_m2&lt;-data_m[data_m$variable!=&quot;MethSNV&quot;,]</span><br><span class="line">p &lt;- ggplot(data_m2, aes(x=xvariable, y=value,color=variable,group=variable)) + stat_smooth(method=&quot;auto&quot;, se=FALSE)</span><br><span class="line">p +labs(title=&quot;&quot;,x=&quot;&quot;, y = &quot;Phased number&quot;)+</span><br><span class="line">  theme_bw() + ## 背景</span><br><span class="line">  theme(panel.grid =element_blank()) + ## 删去网格线</span><br><span class="line">  theme(plot.title = element_text(hjust = 0.5, size = 18, colour = &quot;black&quot;, face = &quot;bold&quot;) ,axis.title.y = </span><br><span class="line">          element_text(size=16,colour = &quot;black&quot;,face = &quot;bold&quot;), </span><br><span class="line">        axis.text.y = element_text(size=10, colour = &quot;black&quot;), </span><br><span class="line">        axis.text.x = element_text(size=16, colour = &quot;black&quot;), legend.position = c(0.85,0.5))</span><br><span class="line">## 箱线图</span><br><span class="line">names(a)=c(&quot;chr&quot;,&quot;coord&quot;,&quot;SNV&quot;, &quot;MethSNV.SNV&quot;, &quot;MethSNV&quot;)</span><br><span class="line">profile_text&lt;-a[c(3,4,5)]</span><br><span class="line">profile_text$xvariable = rownames(profile_text)</span><br><span class="line">data_m &lt;- melt(profile_text, id.vars=c(&quot;xvariable&quot;))</span><br><span class="line">data_m$xvariable &lt;- as.numeric(data_m$xvariable)</span><br><span class="line">data_m2&lt;-data_m[data_m$variable!=&quot;MethSNV&quot;,]</span><br><span class="line">data_m2$value &lt;- pmin(data_m2$value,100)</span><br><span class="line">pdf(&quot;tet.pdf&quot;)</span><br><span class="line">p &lt;- ggplot(data_m2, aes(x=variable, y=value,color=variable,group=variable)) +geom_boxplot()</span><br><span class="line">p</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure><p></p><h3 id="统计meth-HiC合并后的长度分布"><a href="#统计meth-HiC合并后的长度分布" class="headerlink" title="统计meth+HiC合并后的长度分布"></a>统计meth+HiC合并后的长度分布</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">./bshichap HiC_normal/hicsnp_latest_allele_withencodesnp/bowtie_results/bwt2/A549.hic.haplotype WGBS-HAIR/A549.snpwithmr.p.hapcut2.haplo.txt &gt; A549.hapmerge.txt</span><br><span class="line">./bshichap HiC_normal/hicsnp_latest_allele_withencodesnp/bowtie_results/bwt2/A549.hic.haplotype WGBS-HAIR/A549.snpwithmr.n.hapcut2.haplo.txt &gt; A549.hapmerge.n.txt</span><br><span class="line"></span><br><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'BEGIN&#123;a=0;b=0;chr=""&#125;&#123;</span></span><br><span class="line"><span class="string">     if($1~/^=/)&#123;</span></span><br><span class="line"><span class="string">         if(b&gt;a)&#123;print chr,a,b;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">         else&#123;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">     else if(a==0)&#123;</span></span><br><span class="line"><span class="string">         chr=$1;a=$2;b=$2</span></span><br><span class="line"><span class="string">     &#125;else&#123;</span></span><br><span class="line"><span class="string">         if($2&gt;b+3000)&#123;</span></span><br><span class="line"><span class="string">             if(b&gt;a)&#123;print chr,a,b; a=$2;b=$2;chr=$1&#125;</span></span><br><span class="line"><span class="string">         &#125;else if($2&gt;b)&#123;</span></span><br><span class="line"><span class="string">             b=$2</span></span><br><span class="line"><span class="string">         &#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(b&gt;a)&#123;print chr,a,b&#125;&#125;'</span> A549.hapmerge.txt A549.hapmerge.n.txt | sort -k1,1 -k2,2n -k3,3n | bedtools merge | awk <span class="string">'&#123;a+=($3-$2);b++&#125;END&#123;print a,b,a/b&#125;'</span></span><br><span class="line">742297819 1630041 455.386</span><br></pre></td></tr></table></figure><p>统计与repeat的覆盖情况<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'BEGIN&#123;a=0;b=0;chr=""&#125;&#123;</span></span><br><span class="line"><span class="string">     if($1~/^=/)&#123;</span></span><br><span class="line"><span class="string">         if(b&gt;a)&#123;print chr,a,b;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">         else&#123;a=0;b=0;chr="";&#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">     else if(a==0)&#123;</span></span><br><span class="line"><span class="string">         chr=$1;a=$2;b=$2</span></span><br><span class="line"><span class="string">     &#125;else&#123;</span></span><br><span class="line"><span class="string">         if($2&gt;b+3000)&#123;</span></span><br><span class="line"><span class="string">             if(b&gt;a)&#123;print chr,a,b; a=$2;b=$2;chr=$1&#125;</span></span><br><span class="line"><span class="string">         &#125;else if($2&gt;b)&#123;</span></span><br><span class="line"><span class="string">             b=$2</span></span><br><span class="line"><span class="string">         &#125;</span></span><br><span class="line"><span class="string">     &#125;</span></span><br><span class="line"><span class="string">&#125;END&#123;if(b&gt;a)&#123;print chr,a,b&#125;&#125;'</span> A549.hapmerge.txt A549.hapmerge.n.txt | sort -k1,1 -k2,2n -k3,3n | bedtools merge &gt; A549.hapmerge.split3k.txt</span><br><span class="line"></span><br><span class="line">bedtools intersect -a A549.hapmerge.split3k.txt -b ~/practice/Genome/hg38/hg38.fa.out.repeat -wo | awk <span class="string">'&#123;a[$4" "$5" "$6]&#125;END&#123;print length(a)&#125;'</span></span><br><span class="line">2150406</span><br></pre></td></tr></table></figure><p></p><p>统计长度：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">'&#123;if($3-$2&gt;2000)&#123;a++&#125;else if($3-$2&gt;1000)&#123;b++&#125;else if($3-$2&gt;500)&#123;c++&#125;&#125;END&#123;print a,b,c&#125;'</span> A549.hapmerge.split3k.txt</span><br><span class="line">41571 38724 142570</span><br></pre></td></tr></table></figure><p></p><h3 id="A549-单倍型间隔，在HiC中的情况"><a href="#A549-单倍型间隔，在HiC中的情况" class="headerlink" title="A549 单倍型间隔，在HiC中的情况"></a>A549 单倍型间隔，在HiC中的情况</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -h A549.allmerge.sort.bam chr2:89771503-89802405 | samtools sort -o chr2.89771503-89802405.sort.bam -</span><br><span class="line">[10:25:40] qwzhou@login:~/practice/A549_BS/HiC_normal/hicsnp_latest_allele_withencodesnp/bowtie_results/bwt2 :</span><br><span class="line">$ samtools view -h chr2.89771503-89802405.sort.bam |less</span><br><span class="line">[10:26:01] qwzhou@login:~/practice/A549_BS/HiC_normal/hicsnp_latest_allele_withencodesnp/bowtie_results/bwt2 :</span><br><span class="line">$ samtools index chr2.89771503-89802405.sort.bam</span><br><span class="line"><span class="comment">##这个区间不行</span></span><br></pre></td></tr></table></figure><p>区间2:<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">samtools view -h A549.allmerge.sort.bam chr1:125147238-125170518 | samtools sort -n - | samtools view -h | awk <span class="string">'$1~/^@/ || ($7=="=" &amp;&amp; $8&gt;125147238-150 &amp;&amp; $8&lt;125170518+150)'</span> | samtools view -hbS -o chr1.125147238-125170518.sortbyname.bam -</span><br><span class="line"><span class="comment">##这个区间不行</span></span><br></pre></td></tr></table></figure><p></p><p>区间3:<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">samtools view -h A549.allmerge.sort.bam chr17:21850913-26796347 | samtools sort -n - | samtools view -h | awk <span class="string">'$1~/^@/ || ($7=="=" &amp;&amp; $8&gt;21850913-150 &amp;&amp; $8&lt;26796347+150)'</span> | samtools view -hbS -o chr17.21850913-26796347.sortbyname.bam -</span><br><span class="line"><span class="comment">###区间不行</span></span><br></pre></td></tr></table></figure><p></p><p>区间4:<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"> samtools view -h A549.allmerge.sort.bam chr9:43317096-43348509 | samtools sort -n - | samtools view -h | awk <span class="string">'$1~/^@/ || ($7=="=" &amp;&amp; $8&gt;43317096-150 &amp;&amp; $8&lt;43348509)'</span> | samtools view -hbS -o chr9.43317096-43348509.sortbyname.bam -</span><br><span class="line"><span class="comment">###区间不同</span></span><br></pre></td></tr></table></figure><p></p><p>区间5:<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">samtools view -h A549.allmerge.sort.bam chr20:29263493-29295290 | samtools sort -n - | samtools view -h | awk <span class="string">'$1~/^@/ || ($7=="=" &amp;&amp; $8&gt;29263493-150 &amp;&amp; $8&lt;29295290)'</span> | samtools view -hbS -o chr20.29263493-29295290.sortbyname.bam -</span><br><span class="line"><span class="comment">####不行</span></span><br></pre></td></tr></table></figure><p></p><p>区间6<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">samtools view -h A549.allmerge.sort.bam chr21:9035909-9342756 | samtools sort -n - | samtools view -h | awk <span class="string">'$1~/^@/ || ($7=="=" &amp;&amp; $8&gt;9035909-150 &amp;&amp; $8&lt;9342756)'</span> | samtools view -hbS -o chr21.9035909-9342756.sortbyname.bam -</span><br><span class="line"><span class="comment">##不行</span></span><br></pre></td></tr></table></figure><p></p><p>区间7<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">samtools view -h A549.allmerge.sort.bam chr16:34896690-34921521 | samtools sort -n - | samtools view -h | awk <span class="string">'$1~/^@/ || ($7=="=" &amp;&amp; $8&gt;34896690-150 &amp;&amp; $8&lt;34921521)'</span> | samtools view -hbS -o chr16.34896690-34921521.sortbyname.bam -</span><br><span class="line"><span class="comment">##不行</span></span><br></pre></td></tr></table></figure><p></p><p>筛选长的相邻排序：<br></p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk -v OFS=<span class="string">"\t"</span> <span class="string">'&#123;print $3-$2,$1,$2,$3,$4&#125;'</span> A549.hapcut.bed | awk <span class="string">'$1&gt;5000'</span> | sort -k2,2 -k3,3n -k1,1nr |less</span><br></pre></td></tr></table></figure><p></p><p>chr15可以</p><h3 id="A549-SNP数目"><a href="#A549-SNP数目" class="headerlink" title="A549 SNP数目"></a>A549 SNP数目</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">[14:28:03] qwzhou@login:~/practice/A549_BS/WGBS-HAIR :</span><br><span class="line">$ awk <span class="string">'$2==0 || $3==0'</span> A549.onlysnp.hapcut2.haplo.txt | wc -l</span><br><span class="line">866162</span><br><span class="line">[14:44:59] qwzhou@login:~/practice/A549_BS/WGBS-HAIR :</span><br><span class="line">$ awk <span class="string">'ARGIND==1 &amp;&amp; $NF~/0\/1/ &amp;&amp; $1!~/^#/&#123;a[$1" "$2]&#125;ARGIND!=1 &amp;&amp; ($2==0 || $2==1)&#123;if($4" "$5 in a)&#123;d[$4" "$5]&#125;&#125;END&#123;print length(d)&#125;'</span> wgbs.new.replace.mdups.snp.filtered.sort.vcf A549.snpwithmr.p.hapcut2.haplo.txt A549.snpwithmr.n.hapcut2.haplo.txt</span><br><span class="line">1261123</span><br><span class="line">[14:31:23] qwzhou@login:~/practice/A549_BS/WGBS-HAIR :</span><br><span class="line">$ awk <span class="string">'$NF~/0\/1/'</span> wgbs.new.replace.mdups.snp.filtered.sort.vcf |wc -l</span><br><span class="line">1618244</span><br></pre></td></tr></table></figure><p>SNP比例：866162/1618244=53.5%<br>DNA甲基化组装中SNP比例：1261123/1618244=77.9%</p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## A549 合并甲基化单倍型</span></span><br><span class="line">[16:13:39] qwzhou@mn02:~/project/methhap/WGBS/A549 :</span><br><span class="line">$ ../bsmerge A549.snpwithmr.p.hapcut2.haplo.txt A549.snpwithmr.n.hapcut2.haplo.txt &gt; A549.snpwithmr.merge.haplo.txt</span><br></pre></td></tr></table></figure><h2 id="相关文章"><a href="#相关文章" class="headerlink" title="相关文章"></a>相关文章</h2><p>Chen, Z., Hagen, D. E., Ji, T., Elsik, C. G., &amp; Rivera, R. M. (2017). Global misregulation of genes largely uncoupled to DNA methylome epimutations characterizes a congenital overgrowth syndrome. <em>Scientific reports</em>, <em>7</em>(1), 12667. doi:10.1038/s41598-017-13012-z</p><p>值得注意的是，在骨骼肌中，参与成肌细胞增殖和肌管融合的多种途径在LOS胎儿中被错误调节。此外，骨骼肌DNA甲基组的特征表明LOS和对照组之间存在许多局部甲基化差异；然而，只有一小部分差异表达基因（DEG），包括印记基因IGF2R，可能与相邻的差异甲基化区域有关。总之，我们不仅表明非印记基因的错误调节和印记缺失是ART诱导的过度生长综合征的特征，而且证明大多数DEG与DNA甲基体表观突变没有直接关联。</p><p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5477265/" target="_blank" rel="noopener">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5477265/</a></p><p>Hap-ASM</p><p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5320668/" target="_blank" rel="noopener">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5320668/</a></p><p>Vincent, M., Mundbjerg, K., Skou Pedersen, J., Liang, G., Jones, P. A., Ørntoft, T. F., … Wiuf, C. (2017). epiG: statistical inference and profiling of DNA methylation from whole-genome bisulfite sequencing data. <em>Genome biology</em>, <em>18</em>(1), 38. doi:10.1186/s13059-017-1168-4</p><p>DNA甲基化单倍型与核小体占有率</p><p>Cheung, W. A., Shao, X., Morin, A., Siroux, V., Kwan, T., Ge, B., … Grundberg, E. (2017). Functional variation in allelic methylomes underscores a strong genetic contribution and reveals novel epigenetic alterations in the human epigenome. <em>Genome biology</em>, <em>18</em>(1), 50. doi:10.1186/s13059-017-1173-7</p><p>七百多个样本检测ASM与表型相关的基因。</p><p>Marzi, S. J., Meaburn, E. L., Dempster, E. L., Lunnon, K., Paya-Cano, J. L., Smith, R. G., … Mill, J. (2016). Tissue-specific patterns of allelically-skewed DNA methylation. <em>Epigenetics</em>, <em>11</em>(1), 24–35. doi:10.1080/15592294.2015.1127479</p><p>虽然DNA甲基化通常被认为是对称的两个等位基因，但也有一些值得注意的例外。基因组印记和X染色体失活是等位基因特异性甲基化（ASM）的两个被广泛研究的来源，但最近的研究表明，更复杂的模式，即基因型变异可能与顺式结构中等位基因偏斜的DNA甲基化有关。鉴于已知的跨组织和细胞类型的DNA甲基化的异质性，我们研究了人脑的几个区域和来自多个个体的全血中ASM的个体间和个体内变异。与以前的研究一致，我们发现广泛的ASM，在被询问的22万个位点中有4%以上的位点显示出等位基因甲基化的证据。我们确定了ASM侧翼已知的印记区域，并表明ASM位点富含dnase I超敏位点，通常位于中间DNA甲基化的扩展基因组背景中。我们还发现了基因型驱动的ASM的例子，其中一些是组织特异性的。这些发现有助于我们理解组织间差异DNA甲基化的性质，并对复杂疾病的遗传学研究具有重要意义。作为社区的一种资源，所研究的每个组织的ASM模式都可以在一个可搜索的在线数据库中找到：<a href="http://epigenemics.essex.ac.uk/asmbranblood。" target="_blank" rel="noopener">http://epigenemics.essex.ac.uk/asmbranblood。</a></p><p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5373555/" target="_blank" rel="noopener">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5373555/</a></p><p>Li, W., Xu, W., Fu, G., Ma, L., Richards, J., Rao, W., … Song, Q. (2015). High-accuracy haplotype imputation using unphased genotype data as the references. <em>Gene</em>, <em>572</em>(2), 279–284. doi:10.1016/j.gene.2015.07.082</p><p>大量增长的基因组数据集对缺失的数据输入提出了新的挑战，这是一项众所周知的资源需求任务。单倍型需要种族匹配的参考。然而，迄今为止，在世界上大多数人口中没有单倍型参考。我们试图利用现有的非相位基因型（这是啥？）数据作为参考；如果成功，它将覆盖世界上几乎所有的种群。结果表明，我们的hifi软件成功地获得了99.43%的准确率，与无相基因型参考。该方法为突破大数据时代单倍型插补参考可用性有限的瓶颈提供了一种经济有效的解决方案。</p><p>单倍型的重要性</p><p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4254418/" target="_blank" rel="noopener">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4254418/</a></p><p>Glusman, G., Cox, H. C., &amp; Roach, J. C. (2014). Whole-genome haplotyping approaches and genomic medicine. <em>Genome medicine</em>, <em>6</em>(9), 73. doi:10.1186/s13073-014-0073-7</p><p>报道的单倍型而非基因型的基因组信息对于个体化医学将越来越重要。目前的技术产生的二倍体序列数据很少被分解成其组成的单倍型。此外，考虑基因组信息的范例是基于解释基因型而不是单倍型。然而，单倍型在历史上一直是有用的背景下，从群体遗传学到疾病基因绘图的努力。分阶段基因组序列数据的主要方法是分子单倍体、遗传单倍体和基于人群的推断。长读测序技术使分子单倍型更长，而全基因组测序成本的降低使全染色体遗传单倍型测序成为可能。混合方法结合了高吞吐量短读汇编和战略方法，使物理或虚拟分块的读单倍型是使多基因单倍型产生的单个个体。这些技术可以与遗传和种群方法进一步结合。在这里，我们回顾了全基因组单倍体方法的进展，并讨论了单倍体在基因组医学中的重要性。临床应用包括通过识别复合杂合子和分阶段调节变异编码变异来诊断。单倍型比单核苷酸变异体等不太复杂的变异体更具特异性，在预后和诊断、肿瘤分析和移植组织分型中也有应用。未来的进展将包括技术创新，应用标准指标评估单倍型质量，以及开发将单倍型与疾病联系起来的数据库。</p></div><div><div style="text-align:center;color:#ccc;font-size:14px"> ------ The Happy Ending -----</div></div><div><div class="my_post_copyright"><script src="//cdn.bootcss.com/clipboard.js/1.5.10/clipboard.min.js"></script><script src="https://cdn.bootcss.com/jquery/2.0.0/jquery.min.js"></script><script src="https://unpkg.com/sweetalert/dist/sweetalert.min.js"></script><p><span>本文标题:</span><a href="/2019/04/03/project-methhap/">project_methhap</a></p><p><span>文章作者:</span><a href="/" title="访问 Qiangwei 的个人博客">Qiangwei</a></p><p><span>发布时间:</span>2019年04月03日 - 15:04</p><p><span>最后更新:</span>2019年05月11日 - 18:05</p><p><span>原始链接:</span><a href="/2019/04/03/project-methhap/" title="project_methhap">https://zhouqiangwei.github.io/2019/04/03/project-methhap/</a><span class="copy-path" title="点击复制文章链接"><i class="fa fa-clipboard" data-clipboard-text="https://zhouqiangwei.github.io/2019/04/03/project-methhap/" aria-label="复制成功！"></i></span></p><p> 转载请保留原文链接及作者。</p></div><script>var clipboard=new Clipboard(".fa-clipboard");$(".fa-clipboard").click(function(){clipboard.on("success",function(){swal({title:"",text:"复制成功",icon:"success",showConfirmButton:!0})})})</script></div><footer class="post-footer"><div class="post-tags"><a href="/tags/project/" rel="tag"><i class="fa fa-tag"></i> project</a><a href="/tags/methhap/" rel="tag"><i class="fa fa-tag"></i> methhap</a></div><div class="post-nav"><div class="post-nav-next post-nav-item"><a href="/2019/03/19/project-log/" rel="next" title="project_log"><i class="fa fa-chevron-left"></i> project_log</a></div><span class="post-nav-divider"></span><div class="post-nav-prev post-nav-item"> <a href="/2019/04/03/project-alleleweb/" rel="prev" title="project_alleleweb">project_alleleweb<i class="fa fa-chevron-right"></i></a></div></div></footer></div></article><div class="post-spread"></div></div></div><div class="comments" id="comments"></div></div><div class="sidebar-toggle"><div class="sidebar-toggle-line-wrap"><span class="sidebar-toggle-line sidebar-toggle-line-first"></span><span class="sidebar-toggle-line sidebar-toggle-line-middle"></span><span class="sidebar-toggle-line sidebar-toggle-line-last"></span></div></div><aside id="sidebar" class="sidebar"><div class="sidebar-inner"><ul class="sidebar-nav motion-element"><li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap"> 文章目录</li><li class="sidebar-nav-overview" data-target="site-overview-wrap"> 站点概览</li></ul><section class="site-overview-wrap sidebar-panel"><div class="site-overview"><div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person"> <img class="site-author-image" itemprop="image" src="/images/avatar.jpg" alt="Qiangwei"><p class="site-author-name" itemprop="name">Qiangwei</p><p class="site-description motion-element" itemprop="description">顺风不浪 逆风不怂</p></div><nav class="site-state motion-element"><div class="site-state-item site-state-posts"> <a href="/archives/"><span class="site-state-item-count">47</span> <span class="site-state-item-name">日志</span></a></div><div class="site-state-item site-state-categories"> <a href="/categories/index.html"><span class="site-state-item-count">24</span> <span class="site-state-item-name">分类</span></a></div><div class="site-state-item site-state-tags"> <a href="/tags/index.html"><span class="site-state-item-count">59</span> <span class="site-state-item-name">标签</span></a></div></nav><div class="feed-link motion-element"><a href="/atom.xml" rel="alternate"><i class="fa fa-rss"></i> RSS</a></div><div class="links-of-author motion-element"><span class="links-of-author-item"><a href="https://github.com/ZhouQiangwei" target="_blank" title="GitHub"><i class="fa fa-fw fa-github"></i> GitHub</a></span><span class="links-of-author-item"><a href="https://www.zhihu.com/people/qiang-wei-97-1/activities" target="_blank" title="知乎"><i class="fa fa-fw fa-gratipay"></i> 知乎</a></span><span class="links-of-author-item"><a href="mailto:qiangwei.zhou2013@gmail.com" target="_blank" title="E-Mail"><i class="fa fa-fw fa-envelope"></i> E-Mail</a></span></div></div></section><section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active"><div class="post-toc"><div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-3"><a class="nav-link" href="#asm与基因表达"><span class="nav-number">1.</span> <span class="nav-text">asm与基因表达</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#asm的分布"><span class="nav-number">2.</span> <span class="nav-text">asm的分布</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#ASM在基因上的分布"><span class="nav-number">2.1.</span> <span class="nav-text">ASM在基因上的分布</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#ASM在exon上的分布"><span class="nav-number">2.2.</span> <span class="nav-text">ASM在exon上的分布</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#获取intron区域"><span class="nav-number">2.3.</span> <span class="nav-text">获取intron区域</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#ASM在intron上的分布"><span class="nav-number">2.4.</span> <span class="nav-text">ASM在intron上的分布</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#获取基因间区"><span class="nav-number">3.</span> <span class="nav-text">获取基因间区</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#获取基因间区不包含promoter"><span class="nav-number">4.</span> <span class="nav-text">获取基因间区不包含promoter</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#查看asm-site在不同功能元件的分布"><span class="nav-number">5.</span> <span class="nav-text">查看asm site在不同功能元件的分布</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#amrfinder"><span class="nav-number">6.</span> <span class="nav-text">amrfinder</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#prec"><span class="nav-number">6.1.</span> <span class="nav-text">prec</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#A549-Oct4-检测"><span class="nav-number">7.</span> <span class="nav-text">A549 Oct4 检测</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#需要先建立mask基因组索引"><span class="nav-number">7.1.</span> <span class="nav-text">需要先建立mask基因组索引</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#bschip"><span class="nav-number">8.</span> <span class="nav-text">bschip</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#hct116"><span class="nav-number">8.1.</span> <span class="nav-text">hct116</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#PrEC"><span class="nav-number">8.2.</span> <span class="nav-text">PrEC</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#LNCaP"><span class="nav-number">8.3.</span> <span class="nav-text">LNCaP</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#运算"><span class="nav-number">8.4.</span> <span class="nav-text">运算</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#6个细胞系的ASMG-GO分析"><span class="nav-number">9.</span> <span class="nav-text">6个细胞系的ASMG GO分析</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#meth单独的单倍型长度"><span class="nav-number">10.</span> <span class="nav-text">meth单独的单倍型长度</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#A549-覆盖repeat情况"><span class="nav-number">11.</span> <span class="nav-text">A549 覆盖repeat情况</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#统计IMR90、hepG2、K562的单倍型结果"><span class="nav-number">12.</span> <span class="nav-text">统计IMR90、hepG2、K562的单倍型结果</span></a></li></ol><li class="nav-item nav-level-2"><a class="nav-link" href="#hic"><span class="nav-number"></span> <span class="nav-text">+ hic</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#覆盖repeat情况"><span class="nav-number">1.</span> <span class="nav-text">覆盖repeat情况</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#整合成脚本"><span class="nav-number">2.</span> <span class="nav-text">整合成脚本</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#ASM-on-exon-脚本部分"><span class="nav-number">3.</span> <span class="nav-text">ASM on exon 脚本部分</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#工作"><span class="nav-number"></span> <span class="nav-text">工作</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#k562-HepG2-单倍型与WGS比较"><span class="nav-number">1.</span> <span class="nav-text">k562 HepG2 单倍型与WGS比较</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#首先完成WGBS-meth-snv检测单倍型"><span class="nav-number">1.1.</span> <span class="nav-text">首先完成WGBS meth+snv检测单倍型</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#WGS单倍型"><span class="nav-number">1.2.</span> <span class="nav-text">WGS单倍型</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#统计k562、hepG2与WGS-hap重合位点，方便可视化。"><span class="nav-number">1.3.</span> <span class="nav-text">统计k562、hepG2与WGS hap重合位点，方便可视化。</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#hepG2"><span class="nav-number">1.3.1.</span> <span class="nav-text">hepG2</span></a></li></ol></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#拟南芥"><span class="nav-number">2.</span> <span class="nav-text">拟南芥</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#A549-HiC-SNPsplit"><span class="nav-number">2.1.</span> <span class="nav-text">A549 HiC SNPsplit</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#A549-ASM分布"><span class="nav-number">2.2.</span> <span class="nav-text">A549 ASM分布</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#统计单倍型长度"><span class="nav-number">2.3.</span> <span class="nav-text">统计单倍型长度</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#统计单倍型总长度"><span class="nav-number">2.4.</span> <span class="nav-text">统计单倍型总长度</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#统计覆盖在基因组上phased信息"><span class="nav-number">2.5.</span> <span class="nav-text">统计覆盖在基因组上phased信息</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#统计meth-HiC合并后的长度分布"><span class="nav-number">3.</span> <span class="nav-text">统计meth+HiC合并后的长度分布</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#A549-单倍型间隔，在HiC中的情况"><span class="nav-number">4.</span> <span class="nav-text">A549 单倍型间隔，在HiC中的情况</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#A549-SNP数目"><span class="nav-number">5.</span> <span class="nav-text">A549 SNP数目</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#相关文章"><span class="nav-number"></span> <span class="nav-text">相关文章</span></a></li></div></div></section><div id="music163player"><iframe frameborder="no" border="0" marginwidth="0" marginheight="0" width="330" height="86" src="//music.163.com/outchain/player?type=2&id=32364032&auto=0&height=66"></iframe></div></div></aside></div></main><footer id="footer" class="footer"><div class="footer-inner"><script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script><div class="copyright">&copy; 2018 &mdash; <span itemprop="copyrightYear">2019</span><span class="with-love" id="animate"><i class="fa fa-user"></i></span> <span class="author" itemprop="copyrightHolder">Qiangwei</span></div><div class="powered-by"><i class="fa fa-user-md"></i> <span id="busuanzi_container_site_pv">本站总访问量<span id="busuanzi_value_site_pv"></span>次</span> <span id="busuanzi_container_site_uv">| 本站访问<span id="busuanzi_value_site_uv"></span>人次</span></div><div class="theme-info"><div class="powered-by"></div> <span class="post-count">| 博客全站共49.3k字</span></div><div style="display:none"><script src="//s95.cnzz.com/z_stat.php?id=1274106223&web_id=1274106223" language="JavaScript"></script></div></div></footer><div class="back-to-top"><i class="fa fa-arrow-up"></i></div></div><script type="text/javascript">"[object Function]"!==Object.prototype.toString.call(window.Promise)&&(window.Promise=null)</script><script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script><script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script><script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script><script type="text/javascript" src="/js/src/utils.js?v=6.3.0"></script><script type="text/javascript" src="/js/src/motion.js?v=6.3.0"></script><script type="text/javascript" src="/js/src/affix.js?v=6.3.0"></script><script type="text/javascript" src="/js/src/schemes/pisces.js?v=6.3.0"></script><script type="text/javascript" src="/js/src/scrollspy.js?v=6.3.0"></script><script type="text/javascript" src="/js/src/post-details.js?v=6.3.0"></script><script type="text/javascript" src="/js/src/bootstrap.js?v=6.3.0"></script><script src="//cdn1.lncld.net/static/js/3.0.4/av-min.js"></script><script src="//unpkg.com/valine/dist/Valine.min.js"></script><script type="text/javascript">var GUEST=["nick","mail","link"],guest="nick,mail,link";guest=guest.split(",").filter(function(e){return-1<GUEST.indexOf(e)}),new Valine({el:"#comments",verify:!1,notify:!1,appId:"L6lf5Fu6idkKfdhjKT1WmTbQ-gzGzoHsz",appKey:"WIycEuuCj4CnQLrJUlWHvPae",placeholder:"Just go",avatar:"mm",guest_info:guest,pageSize:"10"})</script><script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script><script src="https://cdn1.lncld.net/static/js/av-core-mini-0.6.4.js"></script><script>AV.initialize("L6lf5Fu6idkKfdhjKT1WmTbQ-gzGzoHsz","WIycEuuCj4CnQLrJUlWHvPae")</script><script>function showTime(e){var t=new AV.Query(e),c=[],u=$(".leancloud_visitors");u.each(function(){c.push($(this).attr("id").trim())}),t.containedIn("url",c),t.find().done(function(e){var t=".leancloud-visitors-count";if(0!==e.length){for(var n=0;n<e.length;n++){var o=e[n],i=o.get("url"),s=o.get("time"),r=document.getElementById(i);$(r).find(t).text(s)}for(n=0;n<c.length;n++){i=c[n],r=document.getElementById(i);var l=$(r).find(t);""==l.text()&&l.text(0)}}else u.find(t).text(0)}).fail(function(e,t){console.log("Error: "+t.code+" "+t.message)})}function addCount(i){var e=$(".leancloud_visitors"),s=e.attr("id").trim(),r=e.attr("data-flag-title").trim(),t=new AV.Query(i);t.equalTo("url",s),t.find({success:function(e){if(0<e.length){var t=e[0];t.fetchWhenSave(!0),t.increment("time"),t.save(null,{success:function(e){$(document.getElementById(s)).find(".leancloud-visitors-count").text(e.get("time"))},error:function(e,t){console.log("Failed to save Visitor num, with error message: "+t.message)}})}else{var n=new i,o=new AV.ACL;o.setPublicReadAccess(!0),o.setPublicWriteAccess(!0),n.setACL(o),n.set("title",r),n.set("url",s),n.set("time",1),n.save(null,{success:function(e){$(document.getElementById(s)).find(".leancloud-visitors-count").text(e.get("time"))},error:function(e,t){console.log("Failed to create")}})}},error:function(e){console.log("Error:"+e.code+" "+e.message)}})}$(function(){var e=AV.Object.extend("Counter");1==$(".leancloud_visitors").length?addCount(e):1<$(".post-title-link").length&&showTime(e)})</script></body></html>